---
title: "Single-cell RNA-seq - Clustering for myeloid cells"
---

This method utilizes aggregate-based normalization of data.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Initial setup
### Load libraries
```{r message=FALSE}
suppressPackageStartupMessages({
  library(ggplot2)
  library(Seurat)
  library(tidyverse)
  # library(RCurl)
  # library(cowplot)
  # library(harmony)
  # library(scCustomize)
  # library(org.Hs.eg.db)
})
```
### Update paths
```{r}
obj_dir1 <- file.path(obj_dir,"10_visvader")
fig_dir1 <- file.path(fig_dir,"10_visvader","prep")
```

### Load input libraries
```{r}
# Load Seurat objects
load(file.path(obj_dir1, "seurat_sctv2_3.RData")) ## seurat_sct
load(file.path(obj_dir1, "seurat_sct_sct_myeloidv8_4.RData")) ## seurat_myeloid
load(file.path(obj_dir1, "seurat_sct_sct_macv2_4.RData")) ## seurat_mac
load(file.path(obj_dir1, "seurat_sct_sct_macv2_clean_4.RData")) ## seurat_mac_clean
load(file.path(obj_dir1, "seurat_sct_harmony_lymphv8_clean_4.RData")) ## seurat_lymph_clean
```

### Load biomarkers
```{r}
# Define gene lists
all_gns <- rownames(GetAssayData(seurat_sct))

epi_gns <- c("KRT7", "KRT8", "KRT18", "KRT19", "KRT5", "KRT6B", "KRT14", "KRT17")
fib_gns <- c("COL1A1", "ACTA2", "PDPN")
peri_gns <- c("RGS5")
endo_gns <- c("PECAM1")
mast_gns <- c("MS4A2", "KIT")
dc_gns <- c("FCER1A", "CD1A", "IRF4", "IRF8", "IL3RA", "LAMP3")
mac_gns <- c("CD68", "CD163", "CD14", "LYZ", "FCGR3A")
t_gns <- c("CD3D", "CD2", "TRAC", "TRBC1", "TRDC", "CD4", "FOXP3", "CD8A", "GZMB", "GZMK")
nk_gns <- c("KLRD1", "KLRC1")
b_gns <- c("MS4A1", "IGHM", "JCHAIN")
pl_gns <- c("IGHA1", "IGHA2", "IGHG1", "IGHG3", "IGHG4")

# Create feature plots for markers of interest
lst.mks <- list(epi = epi_gns,
                fib = fib_gns,
                peri = peri_gns,
                endo = endo_gns,
                mac = mac_gns,
                dc = dc_gns,
                mast = mast_gns,
                t = t_gns,
                nk = nk_gns,
                b = b_gns,
                pl = pl_gns)

names(lst.mks) <- c("Cancer","Fibro","Peri","Endo","Mono/Mac","DC","Mast","T","NK","B","Plasma")

mks <- unlist(lst.mks)
mye.mks <- unlist(lst.mks[5:7])
lym.mks <- unlist(lst.mks[8:11])

table(mks %in% all_gns) ## 46 genes are all TRUE

```

## Define breast cancer subtypes

```{r}
sts <- unique(seurat_sct@meta.data[c("orig.ident", "subtype")])
subtypes <- sts$orig.ident
names(subtypes) <- sts$subtype
```

## Update clusters
### Whole population

#### Assign cell types to clusters
```{r}
# Redefine population types based on clusters
tab <- table(seurat_sct$seurat_clusters, seurat_sct$immune_type)
i.lymph <- rownames(tab)[which(tab[, "lymphocyte"] > 0)]
i.mye <- rownames(tab)[which(tab[, "myeloid"] > 0)]

seurat_sct$celltype[as.character(seurat_sct$seurat_clusters) %in% i.lymph] <- "Lymphoid"
seurat_sct$celltype[as.character(seurat_sct$seurat_clusters) %in% i.mye] <- "Myeloid"
seurat_sct$celltype <- factor(seurat_sct$celltype, levels = c("Cancer", "Endothelial", "Fibroblast", "Pericyte", "Lymphoid", "Myeloid"))
```

```{r}
# Update clusters based on cell types
tab1 <- table(seurat_sct$seurat_clusters, seurat_sct$celltype)
i.epi <- rownames(tab1)[which(tab1[, "Cancer"] > 0)]
i.endo <- rownames(tab1)[which(tab1[, "Endothelial"] > 0)]
i.fib <- rownames(tab1)[which(tab1[, "Fibroblast"] > 0)]
i.peri <- rownames(tab1)[which(tab1[, "Pericyte"] > 0)]

i.fib <- c("2", "34")
i.epi <- c("3", "4", "6", "7", "8", "10", "11", "13", "14", "15", "16", "18", "19", "20", "21", "22", "23", 
           "24", "26", "27", "28", "29", "32", "35", "37", "38", "39", "41", "45", "49", "50", "52", "53")

celltype_new <- rep(NA, length(seurat_sct$celltype))
celltype_new[seurat_sct$seurat_clusters %in% i.epi] <- "Cancer"
celltype_new[seurat_sct$seurat_clusters %in% i.fib] <- "Fibroblast"
celltype_new[seurat_sct$seurat_clusters %in% i.endo] <- "Endothelial"
celltype_new[seurat_sct$seurat_clusters %in% i.peri] <- "Pericyte"
celltype_new[seurat_sct$seurat_clusters %in% i.lymph] <- "Lymphoid"
celltype_new[seurat_sct$seurat_clusters %in% i.mye] <- "Myeloid"

seurat_sct$celltype_new <- factor(celltype_new, levels = c("Cancer", "Endothelial", "Fibroblast", "Pericyte", "Lymphoid", "Myeloid"))
```

#### Luminal vs basal in cancer cells
```{r}
new.idx <- c(paste0("epi_",i.epi),
             paste0("fib_",i.fib),
             paste0("peri_",i.peri),
             paste0("endo_",i.endo),
             paste0("mye_",i.mye),
             paste0("lym_",i.lymph))

names(new.idx) <- sub("^.+_","",new.idx)

new.idx.sorted <- new.idx[order(as.numeric(names(new.idx)))] ## new labels (ie celltype_cluster#)
seurat_sct@meta.data[["seurat_clusters_v1"]] <- factor(new.idx.sorted[seurat_sct$seurat_clusters],levels=c(new.idx)) # add them as a factor (levels determins the order in the dotplot)

## Reorder epi clusters based on basal activity 
mean.exp <- AverageExpression(seurat_sct,features=mks,group.by="seurat_clusters_v1")$RNA
mean.exp <- t(as.matrix(mean.exp)[mks,])

mean.exp1 <- cbind(`luminal-krt`=rowSums(mean.exp[,1:5]),
                   `basal-krt`=rowSums(mean.exp[,6:10]),
                   mean.exp[,-(1:10)])
rownames(mean.exp1) <- sub("-","_",rownames(mean.exp1))

## basal activity
basal <- rep(NA,length(levels(seurat_sct$seurat_clusters_v1)))
names(basal) <- levels(seurat_sct$seurat_clusters_v1)
epi_clusters <- grep("epi",names(basal),value=T)
basal[epi_clusters] <- mean.exp1[epi_clusters,"basal-krt"]
seurat_sct@meta.data$basal <- basal[seurat_sct$seurat_clusters_v1]

new.idx.1 <- c(names(sort(basal,decreasing=T)),new.idx[-grep("epi",new.idx)])
names(new.idx.1) <- sub("^.+_","",new.idx.1)
new.idx.sorted.1 <- new.idx.1[order(as.numeric(names(new.idx.1)))] ## new labels (ie celltype_cluster#)
seurat_sct@meta.data[["seurat_clusters_v2"]] <- factor(new.idx.sorted.1[seurat_sct$seurat_clusters],levels=c(new.idx.1)) # add them as a factor (levels determins the order in the dotplot)
```



### Myeloid populations
#### check the number of cells in different seurat objects
```{r}
table(rownames(seurat_myeloid@meta.data) %in% rownames(seurat_sct@meta.data))
table(rownames(seurat_mac@meta.data) %in% rownames(seurat_myeloid@meta.data)) # 18200 TRUE
table(rownames(seurat_mac_clean@meta.data) %in% rownames(seurat_myeloid@meta.data)) ## 16567 TRUE

table(seurat_sct$celltype_new)
```

#### Redefine subtypes
```{r}
## seurat_mac_clean => seurat_mac
mac1_cts <- factor(seurat_mac_clean$celltype_broad)
levels(mac1_cts) <- sub("(.+)_TAM$","Mye_TAM_\\1",levels(mac1_cts))
levels(mac1_cts)[13] <- "Mye_Mac_Tissue-resident"

seurat_mac$celltype_broad <- factor(seurat_mac$celltype_broad)
levels(seurat_mac$celltype_broad) <- c("Endothelial","Mono_Mac","Mono_Mac_DC","Mono_Mac_Epi")
seurat_mac$celltype_mac1 <- as.character(seurat_mac$celltype_broad)
seurat_mac$celltype_mac1[names(mac1_cts)] <- as.character(mac1_cts)
table(seurat_mac$celltype_mac1)

## seurat_mac => seurat_myeloid
mac_cts <- factor(seurat_mac$celltype_mac1)

seurat_myeloid$celltype_broad <- factor(seurat_myeloid$celltype_broad)
levels(seurat_myeloid$celltype_broad) <- c("Mye_DC","Mye_Mast","Mye_Mono_Mac") # remove cancer cells
seurat_myeloid$celltype_mac <- as.character(seurat_myeloid$celltype_broad)
seurat_myeloid$celltype_mac[names(mac_cts)] <- as.character(mac_cts)
uniq.mye <- names(table(seurat_myeloid$celltype_mac))
uniq.mye1 <- rep(c("Endothelial","Dendritic","Mono_Mac","Mast","Mono_Mac",NA,"Mono_Mac"),c(1,2,1,1,10,1,1))
names(uniq.mye1) <- uniq.mye

## seurat_myeloid - umap
seurat_myeloid@meta.data$celltype_v3 <- factor(uniq.mye1[seurat_myeloid$celltype_mac],levels=c("Dendritic","Mono_Mac","Mast"))
seurat_myeloid2 <-subset(seurat_myeloid,cells=names(which(!is.na(seurat_myeloid$celltype_v3))))

```

#### Myeloid UMAP plot

##### [Fig S4D]  myeloid cell types
```{r}
png(file.path(fig_dir1,"umap_myeloid_celltype_v3.png"),width=700,height=700)
p <- DimPlot(seurat_myeloid2, group.by = "celltype_v3", label=FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
    theme(legend.text = element_text(size = 24), legend.key.size=unit(5,"mm")) +
    labs(title = NULL)
print(p)
dev.off()

# seurat_myeloid2@meta.data$subtype <- factor(subtypes[seurat_myeloid2$orig.ident],levels=c("ER+","PR+","HER2+","TNBC"))
```

##### [Fig S4D] BC subtypes
```{r}
seurat_myeloid2@meta.data$subtype <- factor(seurat_myeloid2@meta.data$subtype,levels=c("HR","HER2","TNBC"),labels=c("HR+","HER2+","TNBC"))

p <- DimPlot(seurat_myeloid2, group.by = "subtype", label=FALSE, 
             reduction = "umap",raster=FALSE, pt.size=0.1) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
png(file.path(fig_dir1,"umap_myeloid_subtype.png"),width=700,height=700)
print(p)
dev.off()
```

##### [Fig S4D] Patients
```{r}
png(file.path(fig_dir1,"umap_myeloid_sample.png"),width=700,height=700)
p <- DimPlot(seurat_myeloid2, group.by = "orig.ident", label=FALSE, 
             reduction = "umap",raster=FALSE, pt.size=0.1) +
  labs(title = NULL) +
  NoLegend()
print(p)
dev.off()
```

### Lymphoid populations

#### Load lymphoid cell data
```{r}
lymph1_cts <- factor(seurat_lymph_clean$celltype_broad)
uniq.lymph <- sapply(strsplit(levels(lymph1_cts), "_"), function(x) {
  x1 <- rev(x)[-1]
  x1[1] <- toupper(x1[1])
  if (x1[1] %in% c("CD4", "CD8")) {
    x1 <- c("T", x1)
  }
  paste(c("Lym", x1), collapse = "_")
})
names(uniq.lymph) <- levels(lymph1_cts)
uniq.lymph <- sort(uniq.lymph)[c(8, 9, 10:14, 7, 15:17, 4, 1:3, 5:6)]

seurat_lymph_clean@meta.data$celltype_v1 <- factor(uniq.lymph[as.character(seurat_lymph_clean$celltype_broad)], levels = uniq.lymph)
uniq.lymph1 <- rep(c("T_CD4", "T_CD8", "T_other", "NK", "B", "Plasma"), c(2, 5, 4, 1, 3, 2))
uniq.lymph2 <- rep(c("T", NA, "NK", "B", "Plasma"), c(10, 1, 1, 3, 2))
names(uniq.lymph2) <- names(uniq.lymph1) <- uniq.lymph

seurat_lymph_clean@meta.data$celltype_v2 <- factor(uniq.lymph1[seurat_lymph_clean$celltype_v1], levels = c("T_CD4", "T_CD8", "T_other", "NK", "B", "Plasma"))
seurat_lymph_clean@meta.data$celltype_v3 <- factor(uniq.lymph2[seurat_lymph_clean$celltype_v1], levels = c("T", "NK", "B", "Plasma"))

seurat_lymph_clean1 <- subset(seurat_lymph_clean, cells = names(which(!is.na(seurat_lymph_clean$celltype_v3))))
```
#### Lymphoid UMAP plots

##### [Fig S4C] Lymphocyt subtypes
```{r}
png(file.path(fig_dir1, "umap_lymphoid_celltype_v3.png"), width = 700,height = 700)
p <- DimPlot(seurat_lymph_clean1, group.by = "celltype_v3", label = FALSE, reduction = "umap",raster=FALSE, pt.size=0.1) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
print(p)
dev.off()
```

##### [Fig S4C] BC subtypes
```{r}
seurat_lymph_clean1@meta.data$subtype <- factor(seurat_lymph_clean1@meta.data$subtype,levels=c("HR","HER2","TNBC"),labels=c("HR+","HER2+","TNBC"))

png(file.path(fig_dir1, "umap_lymphoid_subtype_v3.png"), width = 700,height = 700)
p <- DimPlot(seurat_lymph_clean1, group.by = "subtype", label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
print(p)
dev.off()
```

##### [Fig S4C] Patients
```{r}
png(file.path(fig_dir1, "umap_lymphoid_sample_v3.png"), width = 700,height = 700)
p <- DimPlot(seurat_lymph_clean1, group.by = "orig.ident", label = FALSE, reduction = "umap",raster=FALSE, pt.size=0.1) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL) + NoLegend()
print(p)
dev.off()
```

## Update Cell Types
```{r}
# Assign myeloid and lymphoid labels
mye_cts <- factor(seurat_myeloid$celltype_v3)
lym_cts <- factor(seurat_lymph_clean$celltype_v3)

seurat_sct$celltype_1 <- as.character(seurat_sct$celltype_new)
seurat_sct$celltype_1[names(mye_cts)] <- as.character(mye_cts)
seurat_sct$celltype_1[names(lym_cts)] <- as.character(lym_cts)
seurat_sct$celltype_1[seurat_sct$celltype_1 %in% c("Myeloid", "Lymphoid")] <- NA
seurat_sct$celltype_1 <- factor(seurat_sct$celltype_1, levels = c("Cancer", "Fibroblast", "Pericyte", "Endothelial", "Mono_Mac", "Dendritic", "Mast", "T", "NK", "B", "Plasma"))
```

### Subsett seurat object to include only cells with defined cell types
```{r}
valid_cells <- which(!is.na(seurat_sct$celltype_1))

# Create a temporary Seurat object excluding NAs for plotting
uniq_cts <- levels(seurat_sct$celltype_1)
seurat_sct1 <- subset(seurat_sct, cells = names(valid_cells))
rm(seurat_sct)
seurat_sct1$celltype_1 <- factor(seurat_sct1$celltype_1,levels=rev(uniq_cts))
seurat_sct1$celltype_2 <- factor(seurat_sct1$celltype_1,levels=uniq_cts)

```


### Feature plot for marker genes
##### [Fig 2A-left] celltype on UMAP

```{r}
png(file.path(fig_dir1, "umap_wp_celltype.png"), width = 700,height = 700)
p <- DimPlot(seurat_sct1, group.by = "celltype_2", label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
print(p)
dev.off()
```
##### [Fig 2A-left] subtype on UMAP

```{r}
seurat_sct1@meta.data$subtype <- factor(seurat_sct1@meta.data$subtype,levels=c("HR","HER2","TNBC"),labels=c("HR+","HER2+","TNBC"))
# wp@meta.data$subtype <- factor(wp@meta.data$subtype,levels=c("HR","HER2","TNBC"),labels=c("HR+","HER2+","TNBC"))

png(file.path(fig_dir1, "umap_wp_subtype.png"), width = 700,height = 700)
p <- DimPlot(seurat_sct1, group.by = "subtype", label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.05) +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
print(p)
dev.off()
```

##### [Fig 2A-left] subtype on UMAP

```{r}
seurat_sct1@meta.data$orig.ident <- factor(seurat_sct1@meta.data$orig.ident,labels=paste0(1:34))

png(file.path(fig_dir1, "umap_wp_samples.png"), width = 700,height = 700)
p <- DimPlot(seurat_sct1, group.by = "orig.ident", label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.05) +
  theme(legend.text = element_text(size = 16),
        legend.key.size=unit(5,"mm")) +
  labs(title = NULL)
print(p)
dev.off()
```

##### [Fig S4A-right] basal activity on UMAP
```{r}
p <- FeaturePlot(seurat_sct1, feature = "basal", label=FALSE, reduction = "umap",raster=FALSE, pt.size=0.1) +
  scale_color_gradient(low = "blue", high = "red", na.value = "grey") +
  theme(legend.text = element_text(size = 24),
        legend.key.size=unit(10,"mm")) +
  labs(title = NULL)
png(file.path(fig_dir1,"umap_wp_basal_activity.png"),width=700,height=700)
print(p)
dev.off()
```

##### [Fig S4A-right] luminal vs basal score (not)
```{r}
me1 <- as.data.frame(mean.exp1[,1:2]) %>%
  tibble::rownames_to_column("cluster") %>%
  filter(grepl("^epi",cluster)) %>%
  mutate(cluster=sub("epi_","",cluster))

p <- ggplot(me1,aes(x=`luminal-krt`,y=`basal-krt`,label=cluster)) +
  geom_point(aes(color=`basal-krt`)) +
  ggrepel::geom_text_repel() +
  ggtitle("luminal and basal keratin expression per cluster") + 
  scale_color_gradient(low = "blue", high = "red", na.value = "grey") +
  theme_bw()

pdf(file.path(fig_dir1,"cancer_cell_keratin_exp.pdf"),width=5,height=5)
print(p)
dev.off()
```

##### [Fig S4B] gene marker plots - whole population
```{r}
# Loop through markers and generate plots
mks1 <- paste0(unlist(lst.mks)," (",rep(names(lst.mks),sapply(lst.mks,length)),")")
i.wp <- c("KRT8","KRT17","COL1A1","RGS5","PECAM1","CD68","MS4A2","IGHG1","IGHA1","CD3D","KLRC1","MS4A1")
mk.wp <- mks1[sapply(i.wp,function(gn)grep(paste0("^",gn),mks1))]
mk.wp[1:2] <- c("KRT8 (Cancer, luminal)","KRT17 (Cancer, basal)")
names(mk.wp) <- i.wp

for (mk in i.wp) {
  p <- FeaturePlot(seurat_sct1, feature = mk, label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
    theme(
      plot.title=element_text(size=40),
      legend.text = element_text(size = 16),
      legend.key.size=unit(5,"mm")) +
    labs(title = mk.wp[mk])
  png(file.path(fig_dir1, "exp_wp", paste0("umap_wp_", mk, ".png")), width = 700,height = 700)
  print(p)
  dev.off()
}
```

##### [Fig S4C] gene marker plots - lymphocytes
```{r}
i.lym <- c("MS4A1","IGHM","IGHA1","IGHG1","CD3D","CD4","FOXP3","CD8A","KLRC1","GZMB","GZMK")
mk.lym <- mks1[sapply(i.lym,function(gn)grep(paste0("^",gn),mks1))]
names(mk.lym) <- i.lym

for(mk in i.lym){
  p <- FeaturePlot(seurat_lymph_clean1, feature = mk, label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
    theme(
      plot.title=element_text(size=40),
      legend.text = element_text(size = 16),
      legend.key.size=unit(5,"mm")) +
    labs(title = mk.lym[mk])
  png(file.path(fig_dir1,"exp_lym",paste0("umap_lym_",mk,".png")),width=700,height=700)
  print(p)
  dev.off()
}
```

##### [Fig S4D] gene marker plots - myeloid cells
```{r}
i.mye <- c("MS4A2","IRF4","CD14","CD68","CD163")
mk.mye <- mks1[sapply(i.mye,function(gn)grep(paste0("^",gn),mks1))]
names(mk.mye) <- i.mye

for(mk in i.mye){
  p <- FeaturePlot(seurat_myeloid2, feature = mk, label = FALSE, reduction = "umap", raster=FALSE, pt.size=0.1) +
    theme(
      plot.title=element_text(size=40),
      legend.text = element_text(size = 16),
      legend.key.size=unit(5,"mm")) +
    labs(title = mk.mye[mk])  
  png(file.path(fig_dir1,"exp_mye",paste0("umap_mye_",mk,".png")),width=700,height=700)
  print(p)
  dev.off()
}
```

##### [Fig 2B] Dotplot - summary of marker expression per cell type
```{r}
cts_long <- levels(seurat_sct1$celltype_1)
cts_short <- c(cts_long[1:3],rep("T",3),cts_long[7:13])
names(cts_short) <- cts_long

seurat_sct1@meta.data$celltype_short <- factor(cts_short[as.character(seurat_sct1$celltype_2)],levels=unique((cts_short)))

pdf(file.path(fig_dir1,"dotplot_wp_celltype.pdf"),width=12,height=6)
DotPlot(seurat_sct1, features = unique(mks), group.by = "celltype_1") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  xlab("") + ylab("")
dev.off()

table(seurat_sct1$celltype_short)
FeaturePlot(seurat_lymph_clean, feature = "CD8A", label=FALSE, reduction = "umap") # endothelial??
FeaturePlot(seurat_lymph_clean, feature = "CD8B", label=FALSE, reduction = "umap") # endothelial??
# FeaturePlot(seurat_lymph_clean, feature = "nCount_RNA", label=FALSE, reduction = "umap",max.cutoff=10000) # endothelial??

DimPlot(seurat_sct1, group.by = "celltype_2", label=FALSE, reduction = "umap",na.value="white")

## relabeling
DimPlot(seurat_sct, group.by = "celltype_lymph1", label=FALSE, reduction = "umap",na.value="white") + NoLegend()
DimPlot(seurat_sct, group.by = "celltype_1", label=FALSE, reduction = "umap",na.value="white")
DimPlot(seurat_sct1, group.by = "celltype_2", label=FALSE, reduction = "umap",na.value="white")

table(seurat_sct1$celltype_2)

n.count <- as.matrix(table(seurat_sct$celltype_2,seurat_sct$orig.ident))
p <- VlnPlot(seurat_sct1,group.by="celltype_1",features="nCount_RNA",raster=FALSE, pt.size=0.1)
png(file.path(fig_dir1,"nCount_RNA_wp.png"),width=700,height=700)
print(p)
dev.off()

norm.n.count <- apply(n.count,2,function(x)x/sum(x))
barplot(norm.n.count,las=2)

##
pdf(file.path(fig_dir1,"umap_wp_celltype.pdf"),width=7,height=7)
DimPlot(seurat_sct, group.by = "celltype_new", label=FALSE, reduction = "umap",raster=FALSE, pt.size=0.1) +
  labs(title = NULL)
dev.off()

```

## Save Seurat Objects
```{r}
# Save the Seurat objects with updated metadata
saveRDS(wp, file.path(obj_dir1, "visvader_seurat_all_subtypes_v3.rds"))
saveRDS(seurat_lymph_clean1, file.path(obj_dir1, "visvader_seurat_lymph_v3.rds"))
saveRDS(seurat_myeloid2, file.path(obj_dir1, "visvader_seurat_mye_v3.rds"))
```
