---
title: "6. Looking into the correlation in different cancer types"
output: html_notebook
---

## Initial setup

### Load libraries
```{r}
### Load libraries - for WEHI-CCLE
suppressPackageStartupMessages({
  library(dplyr)
  library(heatmap3)
  library(ggplot2)
  # library(ggrepel)
  # library(Seurat)
  library(RColorBrewer)
  # library(lme4)
  library(qvalue)
  library(viridis)
  library(ROCR)
  library(GSVA)
  # library(fgsea)
})
source(file.path(src_dir,"utils","h3.R"))
```

### Set paths
```{r}
obj_dir1 <- file.path(obj_dir,"10_visvader")
fig_dir1 <- file.path(fig_dir,"10_visvader")

fig_dir2 <- file.path(fig_dir,"9_ccle")
obj_dir2 <- file.path(obj_dir,"9_ccle")
```

### Load WEHI data

```{r}
xxx <- load(file=file.path(obj_dir1,"color_schema.rda")) # exp.ct.cols1,st.uniq.cols,mod.cols
x <- load(file=file.path(obj_dir1,"df_gn_exp_tum_102324.rda")) # mod.gns,df.tum.exp
xx <- load(file=file.path(obj_dir1,"mod_activity_visvader_all_102324.rda")) # df.pts,df.ers

# wp <- readRDS(file.path(obj_dir1,"visvader_seurat_all_subtypes_v3.rds"))
# mat <- GetAssayData(wp)
# mat <- mat[rownames(mat) %in% used.all.gns, ]
```

### Load CCLE data
```{r}
x <- load(file.path(obj_dir2,"ccle_RNA-seq.rda")) ## ccle.syms, sub.ccle, ccle.m1.breast
table(sub.ccle$PAM50.mRNA) # 19/54 are Luminal
table(mod.gns %in% ccle.syms) ## all genes found in ccle

ccle.exp <- log2(ccle.m1.breast[match(mod.gns,ccle.syms),]+1)
rownames(ccle.exp) <- mod.gns
med.ccle.exp <- apply(ccle.exp,1,function(x1)quantile(x1,1-(19/54))) # 64.8 percentile - representative expression in CCLE cell lines
```

```{r}
identical(df.tum.exp$genes,names(med.ccle.exp)) # TRUE, order of genes in WEHI and CCLE are identical

# df.tum.exp1 <- df.tum.exp[1:7]
df.tum.exp <- df.tum.exp[1:6] %>% cbind(ccle=med.ccle.exp[df.tum.exp$genes])
```

## [Fig S6X - ROC] Define expressed genes in CCLE by setting threshold

```{r}
exp.pos <- df.tum.exp %>% filter(exp.tis =="Cancer") %>% pull(ccle)
exp.neg <- df.tum.exp %>% filter(exp.tis %in% c("T","NK")) %>% pull(ccle)

predictions <- c(exp.pos,exp.neg)
labels <- rep(c(1,-1),c(length(exp.pos),length(exp.neg)))
pred <- prediction(predictions,labels)
perf <- performance(pred, "sens", "spec")

df.roc <- data.frame(cut = perf@alpha.values[[1]], sens = perf@y.values[[1]], spec = perf@x.values[[1]])
idx <- which.max(df.roc$sens + df.roc$spec)
th <- df.roc$cut[idx]
table(df.tum.exp$ccle > th)
ttl <- paste0(
  "Threshold: ",round(th,3),"\n",
  "FPR:",round(1-df.roc$spec[idx],2),"\n",
  "TPR:",round(df.roc$sens[idx],2))
# th <- threshold1(predictions,labels)

pdf(file.path(fig_dir2,"ROC_curve.pdf"),width=7,height=7)
plot(1-df.roc$spec,
     df.roc$sens,
     xlab="False Positive Rate",
     ylab="True Positive Rate",
     lwd= 3,type="l",
     main= "ROC curve - expression threhold in CCLE")
points(1-perf@x.values[[1]][idx],perf@y.values[[1]][idx],pch=19,col=2)
text(0.25,0.9,ttl,pos=1,col=2)
dev.off()
```

```{r}
df.tum.exp.2 <- df.tum.exp %>%
  mutate(cl.smpls = (exp.tis %in% c("T","NK"))*2 + (exp.tis %in% "Cancer")*1 + 1) %>%
  mutate(cl.smpls = factor(cl.smpls,labels=c("others","Cancer","T/NK"))) %>%
  arrange(order(cl.smpls))

cl.cols <- c(others="grey20",Cancer=2,`T/NK`=4)

dc <- densCols(df.tum.exp.2$tum.exp.1,df.tum.exp.2$ccle,nbin=128,
               colramp =colorRampPalette(blues9[-(1:4)]))

pe <- cor(df.tum.exp.2$tum.exp.1,df.tum.exp.2$ccle,method="pe")

pdf(file.path(fig_dir2,"plot_exp_in_ccle_and_cancer_cells.pdf"),width=7,height=7)
par(mar=c(5,5,3,3))
plot(df.tum.exp.2$tum.exp.1,df.tum.exp.2$ccle,pch=20,cex=1,col=dc,
     xlab="In cancer cells, WEHI (log2)",
     ylab="In cell lines, CCLE (log2)",
     main="Expression of four module genes\nin cancer cells and cell lines")
abline(h=th,lty=2)
abline(h=0,v=0)
text(1.5,2,paste0("PCC=",round(pe,3)))
dev.off()

```

#### [Fig S6B] violin plot - gene expression by cell type
```{r}
df.tum.exp.1 <- df.tum.exp %>%
  group_by(exp.tis) %>%
  summarize_all(~median(ccle),.groups="drop") %>%
  arrange(desc(ccle))
df.tum.exp <- df.tum.exp %>%
  mutate(exp.tis=factor(exp.tis,levels=df.tum.exp.1$exp.tis))

p <- ggplot(df.tum.exp,aes(x=factor(exp.tis),y=ccle)) +
  geom_violin(aes(fill=factor(exp.tis))) +
  geom_jitter(width=0.15,height=0,size=1) +
  scale_fill_manual(values=exp.ct.cols1)+
  geom_point(data=df.tum.exp.1,aes(x=exp.tis,y=tum.exp.1),shape=95,size=10) +
  geom_hline(yintercept=th) +
  labs(title="Module gene expression in BC cell lines",x="celltype",y="Expression in BC cell lines") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),legend.position="none")

set.seed(123)
pdf(file.path(fig_dir2,"violin_ccle_expression_raw_042424.pdf"),width=7,height=7)
print(p)
dev.off()
```

#### [Fig 3B] violin plot - gene expression by modules
```{r}
df.tum.exp.2 <- df.tum.exp %>%
  group_by(module) %>%
  summarize_all(~median(ccle),.groups="drop") %>%
  arrange(desc(ccle))
df.tum.exp.3 <- df.tum.exp %>%
  group_by(module) %>%
  summarize(n.pos = sum(ccle >= th),n = n()) %>%
  mutate(txt = paste0(n.pos,"/",n)) %>%
  mutate(y=9)

df.tum.exp <- df.tum.exp %>% 
  mutate(module=factor(module,levels=names(mods)))

p <- ggplot(df.tum.exp,aes(x=module,y=ccle)) +
  geom_violin(aes(fill=factor(module))) +
  geom_jitter(width=0.15,height=0,size=1) +
  scale_fill_manual(values=mod.cols)+
  geom_point(data=df.tum.exp.2,aes(x=module,y=tum.exp.1),shape=95,size=10) +
  geom_text(data=df.tum.exp.3,aes(x=module,y=y,label=txt),color=2, size=6) +
  geom_hline(yintercept=th) +
  labs(title="Module gene expression in BC cell lines",x="",y="Expression in BC cell lines (log2)") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 12),
        axis.text.y = element_text(size = 12),
        legend.position="none")

set.seed(123)
pdf(file.path(fig_dir2,"violin_ccle_expression_module_042424.pdf"),width=4,height=7)
print(p)
dev.off()
```

## Subset `ccle.exp`

```{r}
df.ccle.ct1 <- df.tum.exp %>% 
  filter(ccle >= th) %>%
  rename(med.ccle = ccle,sym=genes) # 153
ccle.exp1 <- ccle.exp[df.ccle.ct1$sym,]

cell.annotation<- sub.ccle %>% 
  dplyr::select(Cell.line.name,PAM50.mRNA) %>% 
  mutate(PAM50.mRNA = factor(PAM50.mRNA,levels=c("Luminal A","Luminal B","Her2amp","Basal-like"))) %>%
  dplyr::arrange(PAM50.mRNA)

# order of annotations/colors are defined here
ccle.exp.graph <- ccle.exp1[,match(cell.annotation$Cell.line.name,colnames(ccle.exp1))]

groups<- cell.annotation$PAM50.mRNA
```

### save data 

```{r}
save(df.ccle.ct1,ccle.exp1,rc1.sub,file=file.path(obj_dir1,"ccle_expressed_genes_042424.rda"))
xx <- load(file=file.path(obj_dir1,"ccle_expressed_genes_042424.rda"))
```

## Correlation of gene expression (all samples)
```{r}
ccle_cormat.all <- cor(t(ccle.exp.graph), use ="everything",method= "spearman")
ccle_cormat.hr <- cor(t(ccle.exp.graph[,grep("Lumina",cell.annotation$PAM50)]), use ="everything",method= "pearson")
```

### [Fig S6C, 3D] check if they are positively or negatively correlated

```{r}
rc1.sub <- mod.side.cols[df.ccle.ct1$sym,]

hc.all <- heatmap3(ccle_cormat.all,distfun=dist,
          scale="none",
          balanceColor=TRUE,
          ColSideColors=rc1.sub,RowSideColors=rc1.sub)

hcr.all <- hc.all$hcr

m.all <- as.character(cutree(hcr.all,k=5))
old.levs <- unique(m.all[hc.all$rowInd])
new.levs <- c("all1 (ers)","all2 (inc)","all3 (apm,inc)","all4 (apm1)","all5 (apm3)")#paste0("all",1:5)
names(new.levs) <- old.levs
m.all1 <- new.levs[m.all]

##
cols <- colorRampPalette(c("navy", "white", "firebrick3"))(1024)

hc.hr <- heatmap3(ccle_cormat.hr,distfun=dist,
          col=cols,scale="none",
          balanceColor=TRUE,
          ColSideColors=rc1.sub,RowSideColors=rc1.sub)

hcr.hr <- hc.hr$hcr

m.hr <- as.character(cutree(hcr.hr,k=5))
old.levs <- unique(m.hr[hc.hr$rowInd])
new.levs <- c("hr1 (ers)","hr2 (inc)","hr3 (apm1)","hr4 (apm,inc)","hr5 (apm3)")#paste0("hr",1:5)
names(new.levs) <- old.levs
m.hr1 <- new.levs[m.hr]

mems <- df.ccle.ct1 %>%
  mutate(all.mem = m.all1) %>%
  mutate(hr.mem = m.hr1)
rownames(mems) <- c()

table(mems$all,mems$hr)
mems$sym[mems$hr == "hr1 (ers)"]
ers.syms <- (df.ccle.ct1 %>% filter(module=="ERS"))$sym
apmtc12.syms <- (df.ccle.ct1 %>% filter(module %in% c("TNFa/NFkB","APM/TC")))$sym
apmtc23.syms <- (df.ccle.ct1 %>% filter(module %in% c("APM/TC","IFN-I")))$sym
apmtc123.syms <- (df.ccle.ct1 %>% filter(module %in% c("TNFa/NFkB","APM/TC","IFN-I")))$sym

ers.i <- mems %>% filter(sym=="ESR1")
ers.gns <- intersect(mems$sym[which(mems$all == ers.i$all & mems$hr == ers.i$hr)],ers.syms)

apm3.i <- mems %>% filter(sym=="HLA-A")
apm3.gns <- intersect(mems$sym[which(mems$all == apm3.i$all & mems$hr == apm3.i$hr)],apmtc23.syms)

apm1.i <-  mems %>% filter(sym=="AKT3")
apm1.gns <- intersect(mems$sym[which(mems$all == apm1.i$all & mems$hr == apm1.i$hr)],apmtc12.syms)

non.corr.gns <- mems$sym[which(mems$all %in% c("all2 (inc)","all3 (apm,inc)") & mems$hr %in% c("hr2 (inc)","hr4 (apm,inc)"))]
non.corr.gns <- intersect(non.corr.gns,unlist(mods[-1]))

##
cons.gns <- list(ERS=ers.gns, # 18
     `TNFa/NFkB`=apm1.gns, # 17
     `IFN-I`=apm3.gns,
     NON_CORR_IMM = non.corr.gns,
     CORR_IMM = c(apm1.gns,apm3.gns)) # 29

# non.corr <- mems$sym[!mems$sym %in% unlist(cons.gns) & mems$sym %in% unlist(mods[-1])] # 89
# cons.gns <- c(cons.gns,list(CORR_IMM = unlist(cons.gns[1:3]), NON_CORR_IMM=non.corr))
sapply(cons.gns,length)

if(0){
  saveRDS(cons.gns,file.path(obj_dir,"9_ccle","conserved_gns_v4.rds"))
}else{
  cons.gns <- readRDS(file.path(obj_dir,"9_ccle","conserved_gns_v4.rds"))
}

cons.gns.v <- rep(names(cons.gns),sapply(cons.gns,length))
names(cons.gns.v) <- unlist(cons.gns)

cons <- rep("non_exp",nrow(df.ccle.ct1))
names(cons) <- df.ccle.ct1$sym
cons[names(cons.gns.v)] <- cons.gns.v
df.ccle.ct1 <- df.ccle.ct1[1:7] %>% mutate(cons_gns = cons)

saveRDS(df.ccle.ct1,file.path(obj_dir,"10_visvader","df_ccle_ct1_v3.rds"))

```

#### correlation within each cell types in single-cell data?

```{r}
c.all <- ccle_cormat.all
c.hr <- ccle_cormat.hr
diag(c.all) <- diag(c.hr) <- NA
mean.c.all <- sapply(cons.gns,function(x){
  sapply(cons.gns,function(y){
    mean(c.all[x,y],na.rm=T)
  })
})
mean.c.hr <- sapply(cons.gns,function(x){
  sapply(cons.gns,function(y){
    mean(c.hr[x,y],na.rm=T)
  })
})

rng.all <- round(range(mean.c.all)*100)
rng.hr <- round(range(mean.c.hr)*100)
max.rng <- max(abs(range(rng.all,rng.hr)))

# cols1 <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(max.rng * 2 + 1)
cols1 <- colorRampPalette(c("navy", "white", "firebrick3"))(max.rng * 2 + 1)

names(cols1) <- -(max.rng):max.rng

col.all <- cols1[as.character(seq(rng.all[1],rng.all[2]))]
col.hr <- cols1[as.character(seq(rng.hr[1],rng.hr[2]))]

##
# lc <- c(brewer.pal(3,"Set1")[2],brewer.pal(3,"Dark2")[1:2],"white")
lc <- c(brewer.pal(3,"Set1")[2],brewer.pal(3,"Dark2")[1:2],"white")
names(lc) <- names(cons.gns)
lcol <- rep("white",nrow(df.ccle.ct1))
names(lcol) <- df.ccle.ct1$sym
for(i in 1:4){
  nm <- names(cons.gns)[i]
  tmp.gns <- cons.gns[[nm]]
  lcol[tmp.gns] <- lc[nm]
}

rc2.sub <- cbind(rc1.sub,conserved=lcol)

pdf(file=file.path(fig_dir2,"v2","heatmap_ccle_cormat_all_042424.pdf"),height=7,width=7)
hc.all <- heatmap3(ccle_cormat.all,distfun=dist,
          scale="none",main="ALL",
          balanceColor=TRUE,
          ColSideColors=rc2.sub,RowSideColors=rc2.sub)
dev.off()

pdf(file=file.path(fig_dir2,"v2","heatmap_ccle_cormat_hr_042424.pdf"),height=7,width=7)
hc.hr <- heatmap3(ccle_cormat.hr,distfun=dist,
          scale="none",main="HR+",
          balanceColor=TRUE,
          ColSideColors=rc2.sub,RowSideColors=rc2.sub)
dev.off()

write.csv(round(mean.c.all,3),file.path(fig_dir2,"v2","heatmap_mean_ccle_cormat_all_042424.csv"))
write.csv(round(mean.c.hr,3),file.path(fig_dir2,"v2","heatmap_mean_ccle_cormat_hr_042424.csv"))
pdf(file=file.path(fig_dir2,"v2","heatmap_mean_ccle_cormat_all_042424.pdf"),height=7,width=7)
h3(mean.c.all,Colv=NA,Rowv=NA,scale="none",#balanceColor=TRUE,
   margins=c(7,7),main="ALL",col=col.all,
   ColSideColors=cbind(conserved=lc),RowSideColors=cbind(conserved=lc))
dev.off()

pdf(file=file.path(fig_dir2,"v2","heatmap_mean_ccle_cormat_hr_042424.pdf"),height=7,width=7)
h3(mean.c.hr,Colv=NA,Rowv=NA,scale="none",#balanceColor=TRUE,
   margins=c(7,7),main="HR+", col=col.hr,
   ColSideColors=cbind(conserved=lc),RowSideColors=cbind(conserved=lc))
dev.off()
```

## Heatmap of gene expression 
### [Fig S6G] all samples
```{r}
coloursSamples <- factor(groups, levels=c("Luminal A","Luminal B","Her2amp","Basal-like"))
coloursSamples <- colorRampPalette(c("purple","green","hotpink","navyblue"))(length(unique(coloursSamples)))[factor(coloursSamples)]
bc.sub <- cbind(`PAM50`=coloursSamples)

# cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(50)
cols <- viridis(100)

th <- 3
norm.ccle.exp.graph <- t(apply(ccle.exp.graph,1,function(x){
  tmp <- (x - mean(x))/sd(x)
  tmp[tmp > th] <- th
  tmp[tmp < -th] <- -th
  return(tmp)
}))

# rc2.sub <- cbind(rc1.sub[,1:3],conserved=lcol)[,c(2,3,4,1)]
#Plot heatmap for CCLE PAM50
ri <- rev(unlist(cons.gns[1:3]))

pdf(file=file.path(fig_dir2,"v2","ccle_heatmap_all_042424.pdf"),width=7,height=7)
heatmap_ccle<- heatmap3(norm.ccle.exp.graph[ri,],
          col=cols,scale="none",
          Rowv=NA,Colv=NA,
          balanceColor=TRUE,
          cexRow=0.5,
          RowSideColors=rc2.sub[ri,c(1,3,2)],
          ColSideColors= bc.sub)
par(xpd=T)
legend(0.7,1.1,c("Luminal A","Luminal B","Her2amp","Basal-like"),
       fill =c("purple","green","hotpink","navyblue"),border=NA, box.lwd = NA) 
# legend(-.15,.3,c("ERS","IFN int","APM/TC"),
#       fill=c("blue","red","pink"),border=NA, box.lwd = NA)
par(xpd=F)
dev.off()

```

### [Fig 3E] luminal samples
```{r}
groups.sub <- groups[groups %in% c("Luminal A","Luminal B")]
is.used <- groups %in% groups.sub
coloursSamples <- factor(groups.sub, levels=c("Luminal A","Luminal B"))
coloursSamples <- brewer.pal(3,"Pastel1")[factor(coloursSamples)]
bc.sub <- cbind(`PAM50`=coloursSamples)

# cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(50)
cols <- viridis(100)

th <- 3
norm.ccle.exp.graph <- t(apply(ccle.exp.graph[,is.used],1,function(x){
  tmp <- (x - mean(x))/sd(x)
  tmp[tmp > th] <- th
  tmp[tmp < -th] <- -th
  return(tmp)
}))

#Plot heatmap for CCLE PAM50
ri <- rev(unlist(cons.gns[1:3]))

o <- order(colMeans(norm.ccle.exp.graph[cons.gns[[1]],]),decreasing=T)
pdf(file=file.path(fig_dir2,"v2","ccle_heatmap_luminal_042424.pdf"),width=7,height=7)
heatmap_ccle<- heatmap3(norm.ccle.exp.graph[ri,o],
          col=cols,scale="none",
          Rowv=NA,Colv=NA,
          balanceColor=TRUE,
          cexRow=0.5,
          RowSideColors=rc2.sub[ri,c(1,3,2)],
          ColSideColors= bc.sub[o])
par(xpd=T)
legend(0.7,1.1,c("Luminal A","Luminal B"),
       fill =brewer.pal(3,"Pastel1")[1:2],border=NA, box.lwd = NA)
# legend(-.15,.3,c("ERS","IFN int","APM/TC"),
#       fill=c("blue","red","pink"),border=NA, box.lwd = NA)
par(xpd=F)
dev.off()

```

### extend conserved genes
```{r}
tnf.tnf <- intersect(cons.gns$`TNFa/NFkB`,mods$`TNFa/NFkB`)
tnf.apm <- intersect(cons.gns$`TNFa/NFkB`,mods$`APM/TC`)
ifn.ifn <- intersect(cons.gns$`IFN-I`,mods$`IFN-I`)
ifn.apm <- intersect(cons.gns$`IFN-I`,mods$`APM/TC`)

cons.gns.1 <- c(cons.gns[-1],
                list(
                  TNF_sub_1=tnf.tnf,
                  TNF_sub_2=tnf.apm,
                  IFN_sub=ifn.ifn,
                  APM_sub=ifn.apm))

```

## [Fig S6E] Pathway Enrichment (Fisher's test)
```{r}
msigdb <- readRDS(file.path(obj_dir,"5_pathway_analysis","msigdb.rds"))
# msigdb.reg <- readRDS(file.path(obj_dir,"5_pathway_analysis","msigdb_reg.rds"))
```

### subsetting `msigdb`
```{r}
tested.gns <- readRDS(file.path(obj_dir,"1-4_prep","common_tested_gns.rds")) # 15397
tested.ccle.gns <- intersect(tested.gns,ccle.syms) # 15089

ms.gns <- unique(unlist(msigdb))
msigdb.1 <- lapply(msigdb,function(x){
	module1 <- x[x %in% tested.ccle.gns]
	return(module1)
}) # 16437

ge10 <- sapply(msigdb.1,length)>=10 & sapply(msigdb.1,length) <= 500
msigdb.2 <- msigdb.1[ge10] # 11375

all.genes <- intersect(unique(unlist(msigdb.2)),tested.ccle.gns) # 13196
```

### Fisher's exact test
```{r}
sig.gns <- cons.gns.1

# Initialize a list to store the negative log q-values for each element in cons.gns.1
nlqs_list <- list()

# Loop over each element of cons.gns.1 and perform the enrichment analysis
for (set_name in names(cons.gns.1)) {
  cat(set_name,"\n")
  # Perform the parallelized Fisher's exact test for each gene set
  pvals <- parallel::mclapply(msigdb.2, function(x) {
    tab <- table(all.genes %in% cons.gns.1[[set_name]], all.genes %in% x)[c("TRUE", "FALSE"), c("TRUE", "FALSE")]
    fisher.test(tab, alternative = "greater")$p.value
  }, mc.cores = 12)
  
  # Calculate the negative log q-values and store them in the list
  nlqs_list[[set_name]] <- -log10(qvalue(unlist(pvals))$qvalue)
}

nlqs <- do.call(cbind,nlqs_list) %>% 
  as.data.frame() %>%
  tibble::rownames_to_column("gene_set")# %>% filter(gene_set != "HALLMARK_INTERFERON_GAMMA_RESPONSE")

# th <- 3
# if(0){
#   sapply(nlqs[-1],function(x)sum(x>th))
#   
#   nlqs %>% filter(ERS > th & NON_CORR < th) %>% arrange(desc(ERS))
#   
#   module1 <- (nlqs %>% filter(`TNFa/NFkB` > th & `IFN-I` < th & NON_CORR < th) %>% arrange(desc(`TNFa/NFkB`)))$gene_set
#   intersect(msigdb.2[["HALLMARK_COMPLEMENT"]],sig.gns$`TNFa/NFkB`) # HALLMARK_INTERFERON_ALPHA_RESPONSE
#   intersect(msigdb.2[["HALLMARK_COAGULATION"]],sig.gns$`TNFa/NFkB`) # HALLMARK_INTERFERON_ALPHA_RESPONSE
#   
#   head(nlqs %>% filter(`TNFa/NFkB` < th & `IFN-I` > th & NON_CORR < th) %>% arrange(desc(`IFN-I`)),10)
#   intersect(msigdb.2[["HALLMARK_INTERFERON_ALPHA_RESPONSE"]],sig.gns$`IFN-I`) # HALLMARK_INTERFERON_ALPHA_RESPONSE
#   
#   nlqs %>% filter(`APM/TC` > th & NON_CORR > th) %>% arrange(desc(`APM/TC`))
#   head(nlqs %>% filter(`APM/TC` < th & NON_CORR > th) %>% arrange(desc(`APM/TC`)),20)
#   intersect(msigdb.2[["GOBP_REGULATION_OF_T_CELL_ACTIVATION"]],sig.gns$non_corr) # c("GOMF_MHC_PROTEIN_COMPLEX_BINDING","GOBP_REUGLATION_OF_T_CELL_ACTIVATION")
# }
```

### [Fig S6E left] pathways enriched in non-correalted
```{r}
##
th <- 3
pars <- c("correlated","non_correlated")
nlqs.1 <- nlqs %>% mutate(idx= factor(pars[(NON_CORR_IMM > th) + 1],levels=pars))
nlqs.1 %>% filter(idx=="non_correlated") %>% pull(gene_set)

#
pdf(file=file.path(fig_dir2,"v2","fisher_corr_vs_noncorr_v2.pdf"),width=7,height=7)
par(mar=c(5,5,3,3))
plot(nlqs$`CORR_IMM`,nlqs$NON_CORR_IMM,type="n",
     xlab="-log10(Q-value), genes correlated in cell lines",
     ylab="-log10(Q-value), genes expressed but not correlated",
     main="Gene set enrichment (Fisher's test)")
abline(h=c(0,th),v=c(0,th),col=c(1,"grey60"))

cols <- c(1,brewer.pal(3,"Dark2")[3])[nlqs.1$idx]
cexs <- c(.5,1.5)[nlqs.1$idx]
points(nlqs$CORR_IMM,nlqs$NON_CORR_IMM,pch=20,cex=cexs,col=cols)
text(25,7,paste0(sum(nlqs.1$idx=="non_correlated")," gene sets\nrelated to 'non-correlated'"),col=brewer.pal(3,"Dark2")[3])
dev.off()

## export table
nlqs.2 <- nlqs.1 %>% filter(NON_CORR_IMM > th) %>% arrange(desc(NON_CORR_IMM))
write.csv(nlqs.2,file.path(table_dir,"module_non_corr_v3.csv"))
write.csv(nlqs.3.out,file.path(fig_dir2,"v2","fisher_ifn_sub_vs_apm_sub_v2.csv"))

```


```{r}
th2 <- 5
nlqs.3 <- nlqs.1 %>%
  filter(NON_CORR_IMM < th) %>%
  mutate(idx=factor((`TNFa/NFkB` > th2) *2 + (`IFN-I` > th2)*1 + 1)) %>%
  arrange(desc(idx),desc(`TNFa/NFkB`),desc(`IFN-I`))

##
pdf(file=file.path(fig_dir2,"v2","fisher_tnf_vs_ifn_v2.pdf"),width=7,height=7)
par(mar=c(5,5,3,3))
plot(nlqs.3$`TNFa/NFkB`,nlqs.3$`IFN-I`,type="n",
     xlab="-log10(Q-value), TNFa/NFkB intrinsic module",
     ylab="-log10(Q-value), IFN-I intrinsic module",
     main="Gene set enrichment (Fisher's test)")
abline(h=c(0,th2),v=c(0,th2),col=c(1,"grey60"))
cols <- c(1,brewer.pal(3,"Dark2")[1:2],4)[(nlqs.3$`TNFa/NFkB` > th2) +  (nlqs.3$`IFN-I` > th2)*2 + 1]
cexs <- c(.5,1.5,1.5,1.5)[(nlqs.3$`TNFa/NFkB` > th2) +  (nlqs.3$`IFN-I` > th2)*2 + 1]
points(nlqs.3$`TNFa/NFkB`,nlqs.3$`IFN-I`,pch=20,cex=cexs,col=cols)
# text(4,4,"1 pathways\nrelated to 'APMTC1/2'",col=brewer.pal(3,"Dark2")[1])
# text(1.5,15,"20 pathways\nrelated to 'APMTC2/3'",col=brewer.pal(3,"Dark2")[2])
dev.off()

nlqs.4 <- nlqs.3 %>% filter(`IFN-I` > th2) %>%
  mutate(idx2=factor((IFN_sub > th2) * 2 + (APM_sub > th2)+1)) %>%
  arrange(desc(idx2),desc(IFN_sub),desc(APM_sub))
pdf(file=file.path(fig_dir2,"v2","fisher_ifn_sub_vs_apm_sub_v2.pdf"),width=7,height=7)
par(mar=c(5,5,3,3))
plot(nlqs.4$IFN_sub,nlqs.4$APM_sub,type="n",
     xlab="-log10(Q-value), IFN_sub intrinsic module",
     ylab="-log10(Q-value), APM_sub intrinsic module",
     main="Gene set enrichment (Fisher's test)")
abline(h=c(0,th2),v=c(0,th2),col=c(1,"grey60"))
cols <- c(1,brewer.pal(3,"Accent")[c(3,2)],4)[nlqs.4$idx2]
cexs <- c(.5,1,1,1)[nlqs.4$idx2]
points(nlqs.4$IFN_sub,nlqs.4$APM_sub,pch=21,cex=cexs,col=1,bg=cols)
points(nlqs.4$IFN_sub,nlqs.4$APM_sub,pch=20,cex=cexs,col=cols)
# text(4,4,"1 pathways\nrelated to 'APMTC1/2'",col=brewer.pal(3,"Dark2")[1])
# text(1.5,15,"20 pathways\nrelated to 'APMTC2/3'",col=brewer.pal(3,"Dark2")[2])
dev.off()

nlqs.3.out <- nlqs.3 %>% filter(idx!=1) %>%
  arrange(idx,desc(`IFN-I`))

##
write.csv(nlqs.3.out,file.path(fig_dir2,"v2","fisher_ifn_sub_vs_apm_sub_v2.csv"))
```


