---
title: "R Notebook"
output: html_notebook
---

## Initial setup

### Set paths
```{r}
fig_dir1 <- file.path(fig_dir,"11_dfbcc_txn","v2")
obj_dir1 <- file.path(obj_dir,"11_dfbcc_txn")

```

### Load packages
```{r}
suppressPackageStartupMessages({
  library(org.Hs.eg.db)
  library(RColorBrewer)
  library(viridis)
  # library(fgsea)
  library(ggplot2)
  library(ggrepel)
  # library(umap)
  library(GSVA)
  library(heatmap3)
  library(dplyr)
})
```

### Load transcriptome data
```{r}
## load normalized count matrix - log2(x+1) transformation
ldf <- readRDS(file.path(obj_dir1,"txn_32_hr_v2.rds"))
ldf <- ldf[names(ldf)!="NormalBreast"]
```

### Load gene signatures 
```{r}
## MSigDB data
msigdb <- readRDS("/Users/kenichishimada/Dropbox (HMS)/_projects_/B06 03 HR-APM/manuscript/Rdata/5_pathway_analysis/msigdb.rds")
prefix <- sub("_.+","",names(msigdb))
table(prefix)

mod.gns <- readRDS(file=file.path(obj_dir,"1-4_prep","module_genes_final_122923.rds"))#[-2]
mod.gns <- lapply(mod.gns,function(x)x[x != "HLA-H"])

msigdb <- c(msigdb,mod.gns)
uniq.gns <- unique(unlist(mod.gns))

ldf_pca <- prcomp(t(ldf),scale.=FALSE)
vars <- round(ldf_pca$sdev^2/sum(ldf_pca$sdev^2)*100,2)
if(0){
  pdf(file.path(fig_dir,"11_dfbcc_txn","pc_var_all-genes_v2.pdf"),width=7,height=7)
  plot(vars,main="Variance explained by each PC",
       xlab="PC index",ylab="Variance (% total)")
  dev.off()
}
```
#### [used] trim duplicated samples - `ldf` to `ldf.2` (all genes) and `ldf.3` (module genes)
```{r}
smpls <- names(ldf)

## genes that exist both in the count matrix and in MSigDB
uniq.egs <- rownames(ldf)
sum(uniq.gns %in% uniq.egs)# 239
uniq.gns[!uniq.gns %in% uniq.egs] # CXCL5,FCGR1B,HLA-G,HLA-H,GBP6,XCL1,KLRC3,KLRC1,KIR3DL3,KIR3DL1,CSF2,GZMH,IL12A,FASLG,CCL20  

## updated module genes
mod.gns.1 <- lapply(mod.gns,function(gns)gns[gns %in% uniq.egs]) 
used.gns <- unlist(mod.gns.1)

## ldf.2 => summarize BOT5644 and BOT6241
bot5644 <- rowMeans(ldf[grep("BOT5644",names(ldf))])
bot6241 <- rowMeans(ldf[grep("BOT6241",names(ldf))])

ldf.2 <- cbind(ldf[!sub("\\.1","",names(ldf)) %in% c("BOT5644","BOT6241")],BOT5644=bot5644,BOT6241=bot6241)
sorted.smpls <- sort(names(ldf.2))
ldf.2 <- ldf.2[sorted.smpls] # 
ldf.3 <- ldf.2[used.gns,]
x <- cor(t(ldf.3),method="pearson")
```
### [used] correlation between genes, mean correlation per module

#### reorder genes within each module
```{r}
lst.used.gns.1 <- list()
for(nm in names(mod.gns.1)){
  ug <- mod.gns.1[[nm]]
  x.tmp <- x[ug,ug]
  h1 <- hclust(as.dist(1 - cor(t(x.tmp), use = "pa")))
  ug1 <- ug[h1$order]
  lst.used.gns.1[[nm]] <- ug1
}
used.gns.1 <- unlist(lst.used.gns.1)
x1 <- x[used.gns.1,used.gns.1]
```

#### [used] side colors
```{r}
lcol <- readRDS("/Users/kenichishimada/Dropbox (HMS)/_projects_/B06 03 HR-APM/manuscript/Rdata/1-4_prep/sigs_sidecols_clust.rds")
lcol <- lcol[used.gns,]
sc <- lcol[,"subclass"]
sc[sc=="white"] <- "steelblue"
lcol[,"subclass"] <- sc
lcol <- lcol[,c(1,3)]
colnames(lcol) <- c("hallmark","module")

## conserved genes
lc <- c(brewer.pal(3,"Set1")[2],brewer.pal(3,"Dark2")[1:2])
cons.gns <- readRDS(file.path(obj_dir,"9_ccle","conserved_gns_v4.rds"))[1:3]
cons.gns <- lapply(cons.gns,function(x)x[x %in% uniq.egs]) # didn't change
names(lc) <- names(cons.gns)

labs <- rep(names(cons.gns),sapply(cons.gns,length))
lcol1 <- rep("white",nrow(lcol))
names(lcol1) <- rownames(lcol)
lcol1[unlist(cons.gns)] <- lc[labs]
lcol <- cbind(lcol,intrinsic=lcol1)
colnames(lcol)[2] <- "whole-tumor"
```

#### [Fig 4A] heatmap - SCC acrossindividual genes
```{r}
if(1){
  pdf(file.path(fig_dir1,"heatmap_individual_genes_v4.pdf"),width=7,height=7)
  hc.m <- heatmap3(x1,scale="none",#col=cols,
            ColSideColors=lcol,
            balanceColor = TRUE,
            # Colv=NA,Rowv=NA,
            main="29 HR+ samples from DFCI (bulk RNA-seq)")
  dev.off()
  
  x2 <- x
  diag(x2) <- NA
  mean.hrs <- sapply(mod.gns.1,function(x){
    sapply(mod.gns.1,function(y){
      mean(x2[x,y],na.rm=T)
    })
  })

  cols <- colorRampPalette(c("navy", "white", "firebrick3"))(201)
  names(cols) <- -100:100
  rng <- round(range(mean.hrs)*100,0)
  cols1 <- cols[as.character(seq(rng[1],rng[2]))]
  
  ##
  uc <- cbind(subclass=c("steelblue","#7FC97F","#FDC086","#BEAED4"))
  pdf(file.path(fig_dir1,"heatmap_mean_genes_v3.pdf"),width=7,height=7)
  hc.mean.m <- heatmap3(mean.hrs,col=cols1,scale="none",
                   Colv=NA,Rowv=NA,
                   ColSideColors=uc,RowSideColors=uc,
                   balanceColor = FALSE,          
                   main="29 in-house HR+ samples (mean)")
  dev.off()
  write.csv(mean.hrs,file.path(fig_dir1,"heatmap_mean_genes_v3.csv"))
}
```

### [not used]PCA - all genes

```{r}
# smpls <- names(ldf.2)
# 
# ldf_pca <- prcomp(t(ldf.3),scale.=FALSE)
# vars <- round(ldf_pca$sdev^2/sum(ldf_pca$sdev^2)*100,2)
# pdf(file.path(fig_dir1,"pc_var_all-genes_v2.pdf"),width=7,height=7)
# plot(vars,main="Variance explained by each PC",
#      xlab="PC index",ylab="Variance (% total)")
# dev.off()
# 
# cols <- ((1:31) >23)+1
# pchs <- rep(20,length(smpls))#rep(c(20,1,3),c(14,8,1))
# names(pchs) <- names(cols) <- smpls
```

### [used] GSVA of modules
```{r}
gs <- tail(msigdb,4)

# Prepare the ssGSEA parameters using ssgseaParam
ssgsea_params <- ssgseaParam(expr = as.matrix(ldf.2), 
                             geneSets = mod.gns.1, 
                             minSize = 10, 
                             maxSize = 500)

# Run GSVA using the 'ssgsea' method
gsva.gs1 <- gsva(ssgsea_params)

df.gs <- as.data.frame(t(gsva.gs)) %>% 
  tibble::rownames_to_column("bot") %>%
  arrange(desc(ERS)) %>%
  mutate(new=paste0("df",seq(nrow(df.gs))))
```

#### [discontinued] tri.heatmap() [not used for now]
```{r}
if(0){
  tri.heatmap <- function(x,rm.diag=TRUE,mar=c(5,5,3,3),...){
    if(!isSymmetric(x)){
      stop("only symmetric matrix can be handled")
    }
    op <- par()
    par(mar=mar)
        
    n <- nrow(x)
    x[lower.tri(x)] <- NA
    if(rm.diag){
      m <- n-1
      idx <- seq(n)[-1]
      diag(x) <- NA
    }else{
      m <- n
      idx <- seq(n)
    }
    cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(50)
    image(seq(m),seq(m),x[seq(m),rev(seq(n)[idx])],col=cols,axes=F,xlab="",ylab="",zlim=c(-1,1),...)
    axis(1,at=seq(m),rownames(x)[seq(m)],las=2)
    axis(2,at=seq(m),rev(rownames(x)[idx]),las=1)
    for(i in seq(n)){
      for(j in seq(n)){
        text(i,n-j+1,round(x[i,j],2))
      }
    }
    on.exit(suppressWarnings(par(op)))
  }
  
  # rownames(tcga.gs) <- c("ERS","NFkB","APM/TC_ext","APM/TC_int","IFN-res")
  cor.gs <- cor(t(tcga.gs),method="pearson")
  
  pdf(file.path(fig_dir,"11_dfbcc_txn","heatmap_pearson_gs.pdf"),width= 7,height=7)
  tri.heatmap(cor.gs,mar=c(7,7,3,3),main="Perason cor. between sigantures")
  dev.off()
}
```


### [Fig 4B] heatmap - GSVA activity of signature
```{r}
##
act1 <- df.gs[2:5]
rownames(act1) <- df.gs$new
uc <- as.vector(uc)
names(uc) <- names(act1) <- c("ERS","TNFa/NFkB","IFN-I","APM/TC") 

cols <- viridis(100)

##
pdf(file.path(fig_dir1,"module_activity_dfci_bulk.pdf"),width= 7,height=7)
heatmap3(act1,col=cols,
         scale="column",Rowv=NA,Colv=NA,
         cexRow=.8,margin=c(10,12),
         ColSideColors = uc,
         main="Module expression\nper patient\n(bulk RNA-seq)")
dev.off()
```

#### [discontinued] pairwise comparison of module activities
```{r}
if(0){
  ## ers vs apm/tc
  pdf(file.path(fig_dir,"11_dfbcc_txn","ers_vs_apmtc_hr.pdf"),width= 7,height=7)
  ggplot(df.gs,aes(x=ERS,y=`APM/TC`,label=new)) +
    geom_point() + 
    geom_label_repel() +
    theme_bw() +
    ggtitle("Activity of ERS and APMTC2") +
    xlab("ERS activity") +
    ylab("APMTC2 activity")
  dev.off()
  
  ## ers vs apm/tc-ext
  ggplot(df.gs,aes(x=ers,y=apmtc_ext,label=idx)) +
    geom_point(aes(color=isCancer)) +
    geom_text_repel() +
    theme_bw() +
    ggtitle("Activity of ERS and APM/TC_ext") +
    xlab("ERS activity") +
    ylab("APM/TC_ext activity")
  
  
  plot(df.gs$ERS,df.gs$`APM/TC`,main="Activity of ERS and APM/TC")
  plot(df.gs$ERS,df.gs$`APM/TC_ext`,main="Activity of ERS and APM/TC_ext")
  plot(df.gs$ERS,df.gs$`APM/TC_int`,main="Activity of ERS and APM/TC_int")
  plot(df.gs$`APM/TC_int`,df.gs$`APM/TC_ext`,main="Activity of APM/TC_int and APM/TC_ext")
}
```

### [used] save data

```{r}
saveRDS(df.gs,file.path(obj_dir1,"df_pcs_mod_activity_v3.rds"))
# df.gs <- readRDS(file.path(obj_dir1,"df_pcs_mod_activity_v3.rds"))
```



