---
title: "4. Correlations in TCGA and METABRIC"
output: html_notebook
---

## Initial setup

### Load library
```{r}
suppressPackageStartupMessages({
  library(RColorBrewer)
  library(dendextend)
  library(gplots)
  library(dplyr)
  library(mclust)
  library(heatmap3)
  library(limma)
  library(org.Hs.eg.db)
})
```

### Set paths
```{r}
proj_dir="~/Dropbox (HMS)/_projects_/B06 03 HR-APM/"

fig_dir <- file.path(proj_dir,"manuscript","figs")
obj_dir <- file.path(proj_dir,"manuscript","RData")

data_dir <- file.path(proj_dir,"data")
tcga_dir <- file.path(data_dir,"brca_tcga")
metabric_dir <- file.path(data_dir,"brca_metabric")
```

### Load data
```{r}
x <- load(file=file.path(obj_dir,"1-4_prep","gene_signature.rda")) # gene.sig, category, hallmarks
y <- load(file=file.path(obj_dir,"1-4_prep","tcga_processed_v3.rda")) # tcga.ids,tcga.clinical,tcga.syms,tcga.m1,tcga.m3
z <- load(file=file.path(obj_dir,"1-4_prep","metabric_processed_v3.rda")) # meta.ids,meta.clinical,meta.syms,meta.m1,meta.m3
```

<!-- #### for consistency - check if the objects are identical - YES -->
```{r}
if(0){
  ##
  tcga.ids1 <- tcga.ids
  tcga.clinical1 <- tcga.clinical
  tcga.syms1 <- tcga.syms
  tcga.m11 <- tcga.m1
  
  meta.ids1 <- meta.ids
  meta.clinical1 <- meta.clinical
  meta.syms1 <- meta.syms
  meta.m11 <- meta.m1
  
  y1 <- load("/Users/kenichishimada/Dropbox (HMS)/_projects_/B06 03 HR-APM/archive/RData/tcga_RNA-seq.rda")
  z1 <- load("/Users/kenichishimada/Dropbox (HMS)/_projects_/B06 03 HR-APM/archive/RData/metabric.rda")
  
  identical(tcga.ids,tcga.ids1)
  identical(tcga.clinical,tcga.clinical1)
  identical(tcga.syms,tcga.syms1)
  identical(tcga.m1,tcga.m11)
  
  identical(meta.ids,meta.ids1)
  identical(meta.clinical,meta.clinical1)
  identical(meta.syms,meta.syms1)
  identical(meta.m1,meta.m11)
}
```


```{r}
sapply(gene.sig,length)
```


```{r}
ers.genes <- gene.sig$ers
apmtc.genes <- unique(unlist(gene.sig[-1]))

sig.gns <- unique(unlist(gene.sig)) ## 988 genes in either of the three signatures
used.gns <- unique(intersect(unlist(gene.sig),intersect(tcga.syms,meta.syms))) ## 988 -> 778 unique
tested.gns  <- intersect(tcga.syms,meta.syms) # 15397

if(0){
	saveRDS(used.gns,file=file.path(obj_dir,"1-4_prep","common_sig_gns_v3.rds"))
	saveRDS(tested.gns,file=file.path(obj_dir,"1-4_prep","common_tested_gns_v3.rds"))
}
```


## QC: identify tested module genes in both TCGA and METABRIC - not used

```{r}
all.gns <- unique(c(tcga.syms,meta.syms,unlist(gene.sig)))
pdf(file.path(fig_dir,"1_prep_genes","genes_overlap.pdf"),width=7,height=7)
vennDiagram(cbind(
	`TCGA`= all.gns %in% tcga.syms,
	`META`= all.gns %in% meta.syms,
	`GeneSig`= all.gns %in% unlist(gene.sig)),
main="Overlap of genes")
dev.off()
```
### TCGA 
#### Correlation between `used.gns` in `tcga.exp.sub` across`used.ids`
```{r}
table(used.gns %in% tcga.syms)
tcga.exp <- tcga.m3[used.gns,] # 778 genes, 1093 samples

smpl.class <- names(tcga.ids)
tcga.sp <- list()
for(cl in smpl.class){
	used.ids <- tcga.ids[[cl]]
	tcga.exp.sub <- tcga.exp[,used.ids]
	tcga.sp[[cl]] <- cor(t(tcga.exp.sub),method="spearman",use="everything")
}

sapply(tcga.sp,function(x)sum(is.na(x))) ## no missing values
```


### METABRIC
#### subset data in Meta - only `used.gns`
```{r}
table(table(meta.syms[meta.syms %in% used.gns])) ## 8 duplicated, leave it for now (use only the first one)
meta.exp <- meta.m3[used.gns,] # 778 genes, 1904 samples
```

## Correlation between `used.gns` in `meta.exp.sub` across`used.ids`
```{r}
smpl.class <- names(meta.ids)
meta.sp <- list()
for(cl in smpl.class){
	used.ids <- meta.ids[[cl]]
	meta.exp.sub <- meta.exp[,used.ids]
	tmp <- cor(t(meta.exp.sub),method="spearman",use="everything")
  idx <- which(rowSums(is.na(tmp))==nrow(tmp)-1)
  tmp1 <- cor(t(meta.exp.sub[idx,]),t(meta.exp.sub),
              method="spearman",use="pairwise.complete.obs")
  tmp[idx,] <- tmp1
  tmp[,idx] <- t(tmp1)
  
  meta.sp[[cl]] <- tmp
}

sapply(meta.sp,function(x)sum(is.na(x))) ## no missing values
```
### Save data on desk
```{r}
if(0){
  save(tcga.sp,meta.sp,file=file.path(obj_dir,"1-4_prep","tcga_metabric_cors.rda"))
}
```

## QC: transcriptome data

### [Fig S1B] Correlation between hallmark genes

```{r}
tri.heatmap <- function(x,rm.diag=TRUE,mar=c(5,5,3,3),...){
  if(!isSymmetric(tcga.sp1)){
    stop("only symmetric matrix can be handled")
  }
  op <- par()
  par(mar=mar)
      
  n <- nrow(x)
  x[lower.tri(x)] <- NA
  if(rm.diag){
    m <- n-1
    idx <- seq(n)[-1]
    diag(x) <- NA
  }else{
    m <- n
    idx <- seq(n)
  }
  cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(50)
  image(seq(m),seq(m),x[seq(m),rev(seq(n)[idx])],col=cols,axes=F,xlab="",ylab="",zlim=c(-1,1),...)
  axis(1,at=seq(m),rownames(x)[seq(m)],las=2)
  axis(2,at=seq(m),rev(rownames(x)[idx]),las=1)
  for(i in seq(n)){
    for(j in seq(n)){
      text(i,n-j+1,round(x[i,j],2))
    }
  }
  on.exit(suppressWarnings(par(op)))
}
```

```{r}
tcga.sp1 <- tcga.sp$hr
meta.sp1 <- meta.sp$hr

hallmarks <- c("ESR1","CD8A","HLA-A")

tcga.hm <- tcga.sp1[hallmarks,hallmarks]
tcga.hm[!upper.tri(tcga.hm)] <- NA
meta.hm <- meta.sp1[hallmarks,hallmarks]
meta.hm[!upper.tri(meta.hm)] <- NA
```

```{r}
pdf(file.path(fig_dir,"4_tcga_metabric","corr_hallmark_tcga.pdf"),width=5,height=5)
tri.heatmap(tcga.hm,main="TCGA")
dev.off()

pdf(file.path(fig_dir,"4_tcga_metabric","corr_hallmark_meta.pdf"),width=5,height=5)
tri.heatmap(meta.hm,main="METABRIC")
dev.off()
```

### Compare mean expression to look at the quality of the datasets
```{r}
tcga.m1 <- tcga.m1[tested.gns,]
meta.m1 <- meta.m1[tested.gns,]
tcga.m3 <- tcga.m3[tested.gns,]
meta.m3 <- meta.m3[tested.gns,]

identical(rownames(tcga.m1),rownames(meta.m1))
med.t <- rowMeans(tcga.m1,na.rm=T)
med.m <- rowMeans(meta.m1,na.rm=T)

med.ht <- rowMeans(tcga.m1[used.gns,tcga.ids$hr],na.rm=T)
med.hm <- rowMeans(meta.m1[used.gns,meta.ids$hr],na.rm=T)
cor.t <- tcga.sp$hr[used.gns,used.gns]
cor.m <- meta.sp$hr[used.gns,used.gns]
diag(cor.t) <- NA
diag(cor.m) <- NA
med.cor.t <- rowMeans(abs(cor.t),na.rm=T)
med.cor.m <- rowMeans(abs(cor.m),na.rm=T)
med.cors <- list(TCGA=med.cor.t,METABRIC=med.cor.m)
```

### [Fig S1CDE] Comparison between tcga and metabric

```{r}
if(0){
  ## Fig S1C
  pdf(file.path(fig_dir,"4_tcga_metabric","mean_expression_2cohorts.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  cols <- densCols(med.t,med.m,nbin=256)
  plot(med.t,med.m,col=cols,pch=20,cex=.4,
       xlab="Mean log2 expression in TCGA",
       ylab="Mean log2 expression in METABRIC",
       main="Comparison of mean expression for individual genes")
  lo <- lowess(med.t,med.m,f=1/3)
  lines(lo,col=2,lwd=2)
  abline(h=meta.exp.th,lty=2,col=1,lwd=2)
  text(15,meta.exp.th,"Background\nbelow this line",pos=3,cex=1)
  dev.off()
  
  ## Fig S1D
  pdf(file.path(fig_dir,"4_tcga_metabric","mean_corr_2cohorts.pdf"),width=5,height=7)
  p.val <- signif(t.test(med.cors$TCGA,med.cors$METABRIC,alterantive="two.sided")$p.value,2)
  boxplot(med.cors,ylab="Mean absolute spearman corr.",
          main=paste0("Magnitude of correlation within two modules\n",
            "(p-value:",signif(p.val,2),")"))
  dev.off()
  
  ## Fig S1E_left
  rng <- range(med.cors)
  pdf(file.path(fig_dir,"4_tcga_metabric","mean_exp_corr_tcga.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  cols <- densCols(med.ht,med.cor.t,nbin=256)
  plot(med.ht,med.cor.t,col=cols,pch=20,cex=1,ylim=rng,
       xlab="Mean log2 expression in TCGA",
       ylab="Mean absolute correlation in TCGA",
       main="TCGA")
  dev.off()
  
  ## Fig S1E_right
  pdf(file.path(fig_dir,"4_tcga_metabric","mean_exp_corr_metabric.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  cols <- densCols(med.hm,med.cor.m,nbin=256)
  plot(med.hm,med.cor.m,col=cols,pch=20,cex=1,ylim=rng,
       xlab="Mean log2 expression in METABRIC",
       ylab="Mean absolute correlation in METABRIC",
       main="METABRIC")
  abline(v=meta.exp.th,lty=2,col=1,lwd=2)
  text(meta.exp.th,0.27,"Background\nleft to this line",pos=4,cex=1)
  dev.off()
}
```

### Validation of Hallmark genes
#### [Fig S1F] TCGA: plot between `ESR1` and `HLA-A`, `PGR` and `HLA-A`
```{r}
if(0){
  pdf(file.path(fig_dir,"4_tcga_metabric","tcga_esr_hlaa_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(tcga.m3[match(c("ESR1","HLA-A"),tcga.syms),tcga.ids$hr]),pch=20,
       main=paste0(length(tcga.ids$hr)," patients in TCGA\n",
                   paste0("Spearman:",round(tcga.sp$hr["ESR1","HLA-A"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="HLA-A (z-score of log2-exp)") # no null values
  # text(-2,-2,paste0("Spearman:",round(tcga.sp$hr["ESR1","HLA-A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"4_tcga_metabric","tcga_esr_cd8a_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(tcga.m3[match(c("ESR1","CD8A"),tcga.syms),tcga.ids$hr]),pch=20,
       main=paste0(length(tcga.ids$hr)," patients in TCGA\n",
                   paste0("Spearman:",round(tcga.sp$hr["ESR1","CD8A"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="CD8A (z-score of log2-exp)") # no null values
  # text(-2,-2,paste0("Spearman:",round(tcga.sp$hr["ESR1","CD8A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"4_tcga_metabric","tcga_pgr_hlaa_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(tcga.m3[match(c("PGR","HLA-A"),tcga.syms),tcga.ids$hr]),pch=20,
       main=paste0(length(tcga.ids$hr)," patients in TCGA\n",
                   paste0("Spearman:",round(tcga.sp$hr["PGR","HLA-A"],2))),
       xlab="PGR (z-score of log2-exp)",ylab="HLA-A (z-score of log2-exp)") # no null values
  # text(-1,2.5,paste0("Spearman:",round(tcga.sp$hr["PGR","HLA-A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"4_tcga_metabric","tcga_esr_pgr_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(tcga.m3[match(c("ESR1","PGR"),tcga.syms),tcga.ids$hr]),pch=20,
       main=paste0(length(tcga.ids$hr)," patients in TCGA\n",
                   paste0("Spearman:",round(tcga.sp$hr["ESR1","PGR"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="PGR (z-score of log2-exp)") # no null values
  # text(-1.5,1,paste0("Spearman:",round(tcga.sp$hr["ESR1","PGR"],2)))
  dev.off()
}
```

#### [Fig S1F] META: plot between `ESR1` and `HLA-A`, `PGR` and `HLA-A`
```{r}
if(0){
  pdf(file.path(fig_dir,"1_prep_genes","meta_esr_hlaa_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(meta.m3[match(c("ESR1","HLA-A"),meta.syms),meta.ids$hr]),pch=20,
       main=paste0(length(meta.ids$hr)," patients in METABRIC\n",
                   paste0("Spearman:",round(meta.sp$hr["ESR1","HLA-A"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="HLA-A (z-score of log2-exp)") # no null values
  # text(-1,-1,paste0("Spearman:",round(meta.sp$hr["ESR1","HLA-A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"1_prep_genes","meta_esr_cd8a_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(meta.m3[match(c("ESR1","CD8A"),meta.syms),meta.ids$hr]),pch=20,
       main=paste0(length(meta.ids$hr)," patients in METABRIC\n",
                   paste0("Spearman:",round(meta.sp$hr["ESR1","CD8A"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="CD8A (z-score of log2-exp)") # no null values
  # text(-1.5,-1,paste0("Spearman:",round(meta.sp$hr["ESR1","CD8A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"1_prep_genes","meta_pgr_hlaa_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(meta.m3[match(c("PGR","HLA-A"),meta.syms),meta.ids$hr]),pch=20,
       main=paste0(length(meta.ids$hr)," patients in METABRIC\n",
                   paste0("Spearman:",round(meta.sp$hr["PGR","HLA-A"],2))),
       xlab="PGR (z-score of log2-exp)",ylab="HLA-A (z-score of log2-exp)") # no null values
  # text(2.5,-1.3,paste0("Spearman:",round(meta.sp$hr["PGR","HLA-A"],2)))
  dev.off()
  
  pdf(file.path(fig_dir,"1_prep_genes","meta_esr_pgr_expression.pdf"),width=7,height=7)
  par(mar=c(5,5,3,3))
  plot(t(meta.m3[match(c("ESR1","PGR"),meta.syms),meta.ids$hr]),pch=20,
       main=paste0(length(meta.ids$hr)," patients in METABRIC\n",
                   paste0("Spearman:",round(meta.sp$hr["ESR1","PGR"],2))),
       xlab="ESR1 (z-score of log2-exp)",ylab="PGR (z-score of log2-exp)") # no null values
  # text(-1.5,1,paste0("Spearman:",round(meta.sp$hr["ESR1","PGR"],2)))
  dev.off()
}
```

### Cluster genes - create dendrograms in TCGA and METABRIC
```{r}
d.tcga <- as.dist(1-tcga.sp1)
d.meta <-as.dist(1-meta.sp1)
dend1 <- as.dendrogram(hclust(d.tcga,method="average"))
dend2 <- as.dendrogram(hclust(d.meta,method="average"))
```

## Identify modules

### [Fig 1A var] Overlap between each gene module

```{r}
gs1 <- sapply(gene.sig,function(gs){
  used.gns %in% gs  
})

if(0){
  pdf(file.path(fig_dir,"1_prep_genes","overlap_gene_sigs.pdf"),width=5,height=5)
  vennDiagram(gs1,main="overlap of module genes")
  dev.off()
}

gs2 <- cbind(ERS=gs1[,1],APMTC=(gs1[,2]|gs1[,3]))
gs.type <- factor(gs2 %*% c(1,2))
levels(gs.type) <- c("ERS","APMTC","Both")
```

### Heatmap and dendrogram - strating point

### Gene selection in the two cohorts

```{r}
tcga.sp1 <- tcga.sp$hr
meta.sp1 <- meta.sp$hr

cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)

hc.t <- heatmap3(tcga.sp1[used.gns,used.gns],
          col=cols,distfun=dist,scale="none",
          balanceColor=TRUE,
          main="TCGA (778 genes)")
hc.m <- heatmap3(meta.sp1[used.gns,used.gns],
          col=cols,distfun=dist,scale="none",
          balanceColor=TRUE,
          main="METABRIC (778 genes)")
```
### Filter co-expressing genes
#### Gene selection in TCGA
```{r}
hcr.t1 <- hcr.t <- hc.t$hcr
hcr.t1$labels[!hcr.t1$labels %in% hallmarks] <- ""

if(0){
  ## define the thers.thold for cutting the dendrogram
  plot(hcr.t1,lwd=.5,cex=.3)
  abline(h=6,col=3)
  abline(h=12.9,col=3)
}
## first define the apmtc and proliferation
c1.t <- cutree(hcr.t,h=12.9)
cl1.t <- tapply(names(c1.t),c1.t,identity)
# table(c1.t,gs.type) 

cc.t <- intersect(cl1.t[[5]],gene.sig$ers) # proliferation
apmtc.t <- intersect(cl1.t[[1]],c(gene.sig$apm,gene.sig$tc)) # apmtc
# which(c1.t==1 & gs.type=="ERS")

## first define the apmtc and proliferation
c2 <- cutree(hcr.t,h=6)
cl2 <- tapply(names(c2),c2,identity)
idx <- grep("ESR1",cl2)
# table(c2,gs.type) 

ers.t <- intersect(cl2[[idx]],gene.sig$ers)

tcga.sig <- list(ers=ers.t,cc=cc.t,apmtc=apmtc.t)
```

#### Gene selection in METABRIC
```{r}
hcr.m1 <- hcr.m <- hc.m$hcr
hcr.m1$labels[!hcr.m1$labels %in% hallmarks] <- ""

if(0){
  plot(hcr.m1,lwd=.5,cex=.3)
  abline(h=8,col=3)
  abline(h=13,col=3)
}

## first define the apmtc and proliferation
c1.m <- cutree(hcr.m,h=13)
cl1.m <- tapply(names(c1.m),c1.m,identity)
# table(c1.m,gs.type) 

cc.m <- intersect(cl1.m[[3]],gene.sig$ers) # proliferation
apmtc.m <- intersect(cl1.m[[1]],c(gene.sig$apm,gene.sig$tc)) # apmtc

## first define the apmtc and proliferation
c2 <- cutree(hcr.m,h=8)
cl2 <- tapply(names(c2),c2,identity)
idx <- c2["ESR1"]
# table(c2,gs.type) 

ers.m <- intersect(cl2[[idx]],gene.sig$ers)

meta.sig <- list(ers=ers.m,cc=cc.m,apmtc=apmtc.m)
```
### Combine both signatures - select genes that are only positively correated within the module

```{r}
nm.sigs <- c("ers","apmtc","cc")

sigs <- list()
for(sig in nm.sigs){
  tmp1 <- tcga.sig[[sig]]
  tmp2 <- meta.sig[[sig]]
  tmp12 <- tmp1[tmp1 %in% tmp2]
  sigs[[sig]] <- tmp12
}

## ERS module
sp1 <- tcga.sp1[sigs$ers,sigs$ers] 
diag(sp1) <- NA
hist(sp1,breaks=100)
sum(sp1 < 0,na.rm=T) # ERS module is all positively correlated

## APM/TC module
sp2 <- tcga.sp1[sigs$apmtc,sigs$apmtc]
diag(sp2) <- NA
hist(sp2,breaks=100)
sum(sp2 < 0,na.rm=T) # 1992 combination is negative

g2 <- colSums(sp2 < 0,na.rm=T)
m2 <- colMeans(sp2,na.rm=T)
any(m2 < 0) # FALSE - none of the correlation is negative on average
plot(sort(g2,decreasing=T))

## trim positive cor between ers and apmtc
sp12 <- tcga.sp1[sigs$ers,sigs$apmtc] 
hist(sp12,breaks=100)
g12 <- colSums(sp12 > 0,na.rm=T)
m12 <- colMeans(sp12)
any(m12 > 0) ## 6 APM/TC genes are positively correlated with ERS on average

pos <- names(which(m12 >= 0)) ## the six genes to be removed
sigs$apmtc <- sigs$apmtc[!sigs$apmtc %in% pos] # refine sigs-apmtc

##
g12 <- rowSums(sp12 > 0,na.rm=T)/ncol(sp12)
m12 <- rowMeans(sp12)
any(m12 > 0)

##
if(0){
  plot(g12,m12,main="mean correlation of ERS genes",
       xlab="ratio of ERS genes whose correlation against APM/TC are positive",
       ylab="mean correlation against APM/TC")
  abline(h=0,v=0)

  plot(g2,m2)
  abline(h=0,v=0)
  
  hist(sp1,breaks=100)
  hist(sp2,breaks=100)
}
```
### finalized signature
```{r}
sapply(sigs,length)
```
### `cc` signature
```{r}
comm.ers.2 <- sigs$cc
```

### Subdividing the APM/TC module
```{r}
comm.gns1 <- c(sigs$ers,sigs$apmtc)
a <- heatmap3(tcga.sp1[comm.gns1,comm.gns1],col=cols,balanceColor=TRUE,scale="none")

##
hcr1 <- hcr <- a$hcr
hcr1$labels[!hcr1$labels %in% hallmarks] <- ""
if(0){
  plot(hcr1,lwd=.5,cex=.3)
  abline(h=1,col=3)
}

c3.0 <- cutree(hcr,h=1)
c3 <- as.numeric(factor(c3.0,levels=c(1,2,4,3)))
names(c3) <- names(c3.0)

mod.gns <- tapply(names(c3),c3,identity)
names(mod.gns) <- c("ERS","TNFa/NFkB","IFN-I","APM/TC")
```

#### Summarize the panel (not used)
```{r}
tcga.sp2 <- tcga.sp1
meta.sp2 <- meta.sp1

diag(tcga.sp2) <- NA
diag(meta.sp2) <- NA

mean.tcga.sp <- sapply(mod.gns,function(x){
  sapply(mod.gns,function(y){
    mean(tcga.sp2[x,y],na.rm=T)
  })
})

mean.meta.sp <- sapply(mod.gns,function(x){
  sapply(mod.gns,function(y){
    mean(meta.sp2[x,y],na.rm=T)
  })
})
```


## Summary gene selection

### [side_colors] save side colors for heatmap
```{r}
## lcol1: hallmark genes
lcol1 <- rep("white",length(used.gns))
names(lcol1) <- used.gns
lcol1[c("ESR1","CD8A","HLA-A")] <- rep(c("steelblue","indianred"),c(1,2))

## lcol2: initial genesets
gs2 <- cbind(ERS=gs1[,1],APMTC=(gs1[,2]|gs1[,3]))
gs.type <- factor(gs2 %*% c(1,2))
levels(gs.type) <- c("ERS","APMTC","Both")
lcol2 <- c(brewer.pal(3,"Pastel1")[2:1],"white")[as.numeric(gs.type)]

# lcol3: final clusters
lcol3 <- rep("white",length(used.gns))
names(lcol3) <- used.gns
lcol3[sigs$ers] <- "steelblue"
lcol3[sigs$apmtc] <- "indianred"
lcol3[sigs$cc] <- "grey"

## lcol4: APM/TC subclasses
lcol4 <- rep("white",length(used.gns))
names(lcol4) <- used.gns
lcol4[names(c3)] <- c("white",brewer.pal(3,"Accent")[c(1,3,2)])[c3]

lcols <- cbind(hallmark=lcol1,`initial collection`=lcol2,`module`=lcol3,`immune`=lcol4)
rownames(lcols) <- used.gns
```

```{r}
slcol3 <- rep(c("steelblue","indianred"),c(1,3))
slcol4 <- c("white",brewer.pal(3,"Accent")[c(1,3,2)])
slcols <- cbind(`module`=slcol3,`immune`=slcol4)
rownames(slcols) <- names(mod.gns)
```

### [Fig 1B, S2A] Heatmaps - all 778 genes
```{r}
pdf(file.path(fig_dir,"1_prep_genes","heatmap_corr_tcga_all.pdf"),width=15,height=15)
hc.t <- heatmap3(tcga.sp1,
          col=cols,distfun=dist,scale="none",
          balanceColor=TRUE,
          ColSideColors=lcols,
          main="TCGA (778 genes)")
dev.off()

pdf(file.path(fig_dir,"1_prep_genes","heatmap_corr_meta_all.pdf"),width=15,height=15)
hc.m <- heatmap3(meta.sp1,
          col=cols,distfun=dist,scale="none",
          balanceColor=TRUE,
          ColSideColors=lcols,
          main="METABRIC (778 genes)")
dev.off()
```

### [Fig 1D, S2X] mean correlation coefficients

```{r}
idx <- 1:4

cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(201)
n.max.cor.1 <- round(max(abs(mean.tcga.sp))*100)
cols1 <- cols[(-100:100) %in% seq(-n.max.cor.1,n.max.cor.1)]

pdf(file.path(fig_dir,"4_tcga_metabric","heatmap_mean_tcga_mods.pdf"),width=7,height=7)
hc.t <- heatmap3(mean.tcga.sp[idx,idx],Rowv=NA,Colv=NA,
    # dendrogram="none",
    col=cols1,scale="none",distfun=dist,
    ColSideColors=slcols[idx,2:1],
    RowSideColors=slcols[idx,],          
    balanceColor = TRUE,
    margins=c(7,7),
    cexRow=1.5,cexCol=1.5,
    main="TCGA (mean correlation)")
dev.off()

n.max.cor.2 <- round(max(abs(mean.meta.sp))*100)
cols2 <- cols[(-100:100) %in% seq(-n.max.cor.2,n.max.cor.2)]

pdf(file.path(fig_dir,"4_tcga_metabric","heatmap_mean_meta_mods.pdf"),width=7,height=7)
hc.m <- heatmap3(mean.meta.sp[idx,idx],Rowv=NA,Colv=NA,
          # dendrogram="none",
          col=cols2,scale="none",distfun=dist,
          ColSideColors=slcols[idx,2:1],
          RowSideColors=slcols[idx,],          
          balanceColor = TRUE,          
          margins=c(7,7),       
          cexRow=1.5,cexCol=1.5,  
          main="METABRIC (mean correlation)")
dev.off()
```

## Save data
### write.csv
```{r}
x <- list(
  ers=data.frame(geneid=unlist(mget(sort(mod.gns$ERS),org.Hs.egSYMBOL2EG,ifnotfound=NA)),symbol=sort(mod.gns$ERS)),
  `TNFa_NFkB`=data.frame(geneid=sapply(mget(sort(mod.gns[["TNFa/NFkB"]]),org.Hs.egSYMBOL2EG,ifnotfound=NA),paste,collapse=";"),
                    symbol=sort(mod.gns[["TNFa/NFkB"]])),
  `IFN-I`=data.frame(geneid=sapply(mget(sort(mod.gns[["IFN-I"]]),org.Hs.egSYMBOL2EG,ifnotfound=NA),paste,collapse=";"),
                    symbol=sort(mod.gns[["IFN-I"]])),  
  `APM_TC`=data.frame(geneid=sapply(mget(sort(mod.gns[["APM/TC"]]),org.Hs.egSYMBOL2EG,ifnotfound=NA),paste,collapse=";"),
                    symbol=sort(mod.gns[["APM/TC"]])),
  prolif=data.frame(geneid=sapply(mget(sort(sigs$cc),org.Hs.egSYMBOL2EG,ifnotfound=NA),paste,collapse=";"),
                    symbol=sort(sigs$cc)))
wb <- openxlsx::buildWorkbook(x)

openxlsx::saveWorkbook(wb,file.path(fig_dir,"Table S2 gene_modules_122923.xlsx"))
```
### as R objects

###
```{r}
if(0){
  saveRDS(sigs,file=file.path(obj_dir,"1-4_prep","modules3_ers_apmtc_cc.rds")) ## ers, apmtc, cc
  sapply(sigs,length)
  # ers apmtc    cc 
  #  29   225   101 
  saveRDS(mod.gns,file=file.path(obj_dir,"1-4_prep","module_genes_final_122923.rds")) ## ers, apmtc, cc
  sapply(mod.gns,length)
  # ERS APM/TC1 APM/TC2 APM/TC3 
  #  29      53     143      29   
  save(lcols,slcols,file=file.path(obj_dir,"1-4_prep","heatmap_sidecols.rda"))
}
```

## Conclusion
`module.gns.rds` is the list of genes used for the rest of the analysis. Since clustering result changes depending on the version of the mclust used, I will stick with this gene set for the rest of the analysis. The old file is copied to the recent `obj_dir` on 9/3/2023 (and for the paper).

