---
title: "6. Looking into the correlation in different cancer types"
output: html_notebook
---

## Initial setup

### Load libraries
```{r}
suppressPackageStartupMessages({
  library(org.Hs.eg.db)
  library(qvalue)
  library(RColorBrewer)
  library(dendextend)
  library(gplots)
  library(dplyr)
  library(ggplot2)
  library(ggsignif)
  library(ggpubr)
  library(rstatix)
  library(gtools)
})
```

### Set paths
```{r}
obj_dir <- "~/Dropbox (HMS)/_projects_/B06 03 HR-APM/data/RData"
fig_dir <- "~/Dropbox (HMS)/_projects_/B06 03 HR-APM/figs"
```

### Load data
```{r}
load(file.path(obj_dir,"ccle_RNA-seq.rda"))
load(file.path(obj_dir,"common_signatures.rda"))
```

## Subset `ccle.exp`

```{r}
comm.gns <- c(comm.esr,comm.apm.tc)

#Filter for 163 ERS an APM/TC panel genes
ccle.exp <- log2(ccle.m1.breast[match(comm.gns,ccle.syms),]+1)
rownames(ccle.exp) <- comm.gns

# ks: check which genes are not expressed - 161, these are removed
sds <- apply(ccle.exp,1,sd)
which(sds==0) ## IFI30, IFNG are not expressed by any cells, thus removed
ccle.exp <- ccle.exp[sds>0,] ## 161 genes, 54 cell lines
comm.gns <- comm.gns[sds>0] # 161 genes

#set col colors
side.col.annotation<- sub.ccle %>% select(Cell.line.name,PAM50.mRNA) %>% arrange(PAM50.mRNA,Cell.line.name)
unique(side.col.annotation$PAM50.mRNA)

# order of annotations/colors are defined here
ccle.exp.graph <- ccle.exp[,match(side.col.annotation$Cell.line.name,colnames(ccle.exp))]

groups<- side.col.annotation$PAM50.mRNA
```

## Heatmap of gene expression (all samples)
```{r}
coloursSamples <- factor(groups, levels=c("Basal-like","Her2amp","Luminal A","Luminal B"))
coloursSamples <- colorRampPalette(c("navyblue","hotpink","purple","green"))(length(unique(coloursSamples)))[factor(coloursSamples)]

lcol <- brewer.pal(3,"Set1")[(comm.gns %in% comm.esr)+1]

cols <- colorRampPalette(rev(brewer.pal(11,"RdBu")))(50)
hc <- function(x)hclust(x,method="complete")

#Plot heatmap for CCLE PAM50
# ks: use log-transformed exppression for heatmap
pdf(file=file.path(fig_dir,"CCLE_heatmap_logexp.pdf"))
heatmap_ccle<- heatmap.2(ccle.exp.graph, trace="none",dendrogram='none',Rowv=FALSE,Colv=FALSE, 
                         col=cols,
                         RowSideColors=lcol,ColSideColors= coloursSamples,
                         hclustfun=hc,cexRow = 0.2,cexCol = 0.2, scale = "row")
legend("topright",c("Basal-like","Her2amp","Luminal A","Luminal B"),fill =c("navyblue","hotpink","purple","green"),border=NA, box.lwd = 0) 
dev.off()
```

## Correlation of gene expression (all samples)
```{r}
##total 161 genes, removed IFI30, and IFNG for cor matrix as not expression was detected.
ccle.exp.graph.cor<- ccle.exp.graph
ccle_cormat<- cor(t(ccle.exp.graph.cor), use ="everything",method= "spearman")
#set ERS, APM/TC gene colors
lcol.cor <- brewer.pal(3,"Set1")[(colnames(ccle_cormat) %in% comm.esr)+1]

pdf(file=file.path("ccle_scmatrix_2021.08.02.pdf"))
hc1 <- heatmap.2(ccle_cormat, trace="none",#dendrogram='none',Rowv=FALSE,Colv=FALSE, 
          RowSideColors=lcol.cor,ColSideColors=lcol.cor,hclustfun=hc,
          col=cols,cexRow = 0.2,cexCol = 0.2) 
dev.off()
```

## Clustering genes
```{r}
gs <- rep(c("ESR","TC"),c(length(comm.esr),length(comm.apm.tc)))
gs <- gs[!c(comm.esr,comm.apm.tc) %in% c("IFI30","IFNG")]

# cut the dendrogram to identify 4 gene clusters
cl.hc <- factor(cutree(hc1$colDendrogram,k=4))
table(hclust=cl.hc,genesig=gs)
```
### Heatmap with gene clusters
```{r}
lcol2.cor <- brewer.pal(6,"Set1")[as.numeric(cl.hc)] ## cl.hc

pdf(file="ccle_cormat_cutree-k4.pdf")
heatmap.2(ccle_cormat, trace="none",#dendrogram='none',Rowv=FALSE,Colv=FALSE, 
          RowSideColors=lcol2.cor,ColSideColors=lcol2.cor,hclustfun=hc,
          col=cols,cexRow = 0.2,cexCol = 0.2) 
dev.off()

if(0){
  ## k-means (via GMM) - I decided not to use this algorithm
  ## because the clusters looked drastically different if IFI30 and IFNG were removed
  ## i.e. the clusters are not robust

  library(mclust)
  bics <- mclustBIC(ccle.exp,G=1:50,modelNames=c("EII"))
  plot(bics) # G=4 looks acceptable - not if the two genes were removed at the beginning
  m <- Mclust(ccle.exp,G=4,modelNames="EII",x=bics)
  
  cl.km <- factor(m$classification) # 161 genes

  table(cluster=cl.km,genesig=gs)
  
  length(cl.km)
}
```

## Heatmap of mean of correlations
```{r}
diag(ccle_cormat) <- NA
mean.cor <- sapply(levels(cl.hc),function(x){
  sapply(levels(cl.hc),function(y){
    round(mean(ccle_cormat[cl.hc==x,cl.hc==y],na.rm=T),3)
  })
})

uniq.lcol <- brewer.pal(4,"Set1")
steps <- seq(-1,1,length=50)
d <- 50-which.min(abs(max(abs(mean.cor))-steps))
sub.cols <- cols[(d+1):(50-d)]
pdf(file="mean_ccle_cormat_4clusters.pdf")
heatmap.2(mean.cor, trace="none",#dendrogram='none',Rowv=FALSE,Colv=FALSE, 
          RowSideColors=uniq.lcol,ColSideColors=uniq.lcol,hclustfun=hc,
          col=sub.cols,cexRow = 1,cexCol = 1,
          main="mean correlation coef.\nbetween clusters") 
dev.off()
```

## Mean correlation coefficients among 4 groups
```{r}
mean.cor[4:1,1:4]
## here, we decided to divide genes into three groups
gns.hc <- tapply(names(cl.hc),cl.hc,identity)

# gns.hc == 1: ERS signature
# gns.hc == 2: extrinsic signature (not correlated with either ERS or APM/TC)
# gns.hc == 3: intrinsic signature (strongly correlated with APM/TC and neg. correlated with ERS)
# gns.hc == 4: intermediate/indeterminant

int.sig <- gns.hc[[3]]
ext.sig <- gns.hc[[2]]

hc.in <- comm.apm.tc[comm.apm.tc %in% int.sig]
hc.ex <- comm.apm.tc[comm.apm.tc %in% ext.sig]
hc.others <- comm.apm.tc[!comm.apm.tc %in% c(hc.in,hc.ex)]

sig.gns <- list(int=hc.in,ext=hc.ex,others=hc.others)
sapply(sig.gns,length)
```

## Pathway enrichment 
### Subset `msigdb`
Looking at the pathways enriched in intrinsic and extrinsic sigs.
```{r}
load(file.path(obj_dir,"msigdb_v6.1.rda"))

keggs <- msigdb[grep("^KEGG_",names(msigdb))]
gos <- msigdb[grep("^GO_",names(msigdb))]

tested.gns <- readRDS(file.path(obj_dir,"common_tested_gns.rds"))

##
kg <- msigdb[sub("_.+","",names(msigdb)) %in% c("HALLMARK","GO","KEGG","REACTOME","BIOCARTA")]
barplot(sort(table(sub("_.+","",names(kg)))),las=2)

kg.syms <- unlist(mget(unique(unlist(kg)),org.Hs.egSYMBOL,ifnotfound=NA))
all.genes <- intersect(kg.syms,tested.gns) # 13318
tab <- as.matrix(sort(table(sub("_.+","",names(kg))),decreasing=T))
colnames(tab)[1] <- "# gene sets"

kg.syms <- lapply(kg,function(x){
  s <- unlist(mget(x,org.Hs.egSYMBOL,ifnotfound=NA))
  s <- s[!is.na(s)]
  s[s %in% all.genes]
})

ge20 <- sapply(kg.syms,length)>=20 & sapply(kg.syms,length) <= 500
kg.syms <- kg.syms[ge20]

all.genes <- intersect(unlist(kg.syms),tested.gns) # 12708 (all pathways)
```

### Fisher's exact test
```{r}
pvals.int <- parallel::mclapply(kg.syms,function(x){
  tab <- table(all.genes %in% sig.gns$int, all.genes %in% x)[c("TRUE","FALSE"),c("TRUE","FALSE")]
  pval <- (fisher.test(tab,alternative="greater")$p.value)
},mc.cores=6)

pvals.ext <- parallel::mclapply(kg.syms,function(x){
  tab <- table(all.genes %in% sig.gns$ext, all.genes %in% x)[c("TRUE","FALSE"),c("TRUE","FALSE")]
  pval <- (fisher.test(tab,alternative="greater")$p.value)
},mc.cores=6)

pvals.oth <- parallel::mclapply(kg.syms,function(x){
  tab <- table(all.genes %in% sig.gns$oth, all.genes %in% x)[c("TRUE","FALSE"),c("TRUE","FALSE")]
  pval <- (fisher.test(tab,alternative="greater")$p.value)
},mc.cores=6)

##
nlqs.int <- -log10(qvalue(unlist(pvals.int))$qvalue)
nlqs.ext <- -log10(qvalue(unlist(pvals.ext))$qvalue)

##
rng <- range(c(nlqs.ext,nlqs.int))
```
## Graphical summary
### Scatter plot, extrinsic vs intrinsic enrichment
```{r}
pdf("pathway_int-ext.pdf")
#idx <- which(nlqs.hc.in > 30 & predict(lm1) > nlqs.hc.ex ) 
in.idx <- which(nlqs.int >= th & nlqs.ext < th)
ex.idx <- which(nlqs.int < th & nlqs.ext >= th)
bo.idx <- which(nlqs.int >= th & nlqs.ext >= th)

cols <- rep(1,length(nlqs.int))
cexs <- rep(1,length(nlqs.int))
cols[bo.idx] <- 2
cols[in.idx] <- 3
cols[ex.idx] <- 4
cexs[c(bo.idx,in.idx,ex.idx)] <- 2
par(mar=c(5,5,3,3))
plot(nlqs.int,nlqs.ext,pch=20,xlim=rng,ylim=rng,col=cols,cex=cexs,
     xlab="Pathways overrepresented by Intrinsic APM genes",
     ylab="Pathways overrepresented by Extrinsic APM genes",
     main="Pathways associated with Intrinsic and Extrinsic genes")

th <- 10
abline(a=0,b=1,lty=1)
abline(a=0,b=1,lty=1)
abline(h=th,v=th)

points(nlqs.int[bo.idx],nlqs.ext[bo.idx],pch=20,col=2)
points(nlqs.int[in.idx],nlqs.ext[in.idx],pch=20,col=3)
points(nlqs.int[ex.idx],nlqs.ext[ex.idx],pch=20,col=4)
dev.off()
```

### Barplots of significant pathways
```{r}
bar.rng <- c(0,max(nlqs.ext,nlqs.int))
pdf("intrinsic_bar.pdf")
par(mar=c(5,20,5,3))
barplot(sort(nlqs.int[nlqs.int>10]),horiz=T,col="steelblue",xlim=bar.rng,
        las=1,xlab="-log10(P-value, adusted)",
        main="Pathways enriched in\nAPM/TC intrinsic gene panel",cex.names=.7)
dev.off()
##
pdf("extrinsic_bar.pdf")
par(mar=c(5,20,5,3))
barplot(sort(nlqs.ext[nlqs.ext>10]),horiz=T,col="steelblue",xlim=bar.rng,
        las=1,xlab="-log10(P-value, adusted)",
        main="Pathways enriched in\nAPM/TC extrinsic gene panel",cex.names=.7)
dev.off()
```
